MSC3575: Sliding Sync (aka Sync v3) (#3575)
matrix-org/matrix-spec-proposals: proposals/3575-sync.md

        "is_encrypted": true,
        // Flag which only returns rooms the user is currently invited to. If unset, both invited
        // and joined rooms are returned. If false, no invited rooms are returned. If true, only
        // invited rooms are returned.
        "is_invite": true,
        // Flag which only returns rooms which have an `m.room.tombstone` state event. If unset,
        // both tombstoned and un-tombstoned rooms are returned. If false, only un-tombstoned rooms
        // are returned. If true, only tombstoned rooms are returned.
        "is_tombstoned": true,
        // If specified, only rooms where the `m.room.create` event has a `type` matching one
        // of the strings in this array will be returned. If this field is unset, all rooms are
        // returned regardless of type. This can be used to get the initial set of spaces for an account.
        "room_types": [ "m.space" ],
        // Same as "room_types" but inverted. This can be used to filter out spaces from the room list.
        "not_room_types": [ "m.space" ],
        // Filter the room name. Case-insensitive partial matching e.g 'foo' matches 'abFooab'.

>         // The term 'like' is inspired by SQL 'LIKE', and the text here is similar to '%foo%'.

Reviewer: As I noticed in [the ruma impl PR](https://github.com/ruma/ruma/pull/999#discussion_r886774323), this is unclear, and potentially misleading;

> This also implies that use of `%` must be in the string, or else it'll not match suffixed or prefixed room names.
> 
> From what i'm reading, this is a simple substring server search.

---

Suggestions; name it `partial_room_name`, drop the mention of SQL, and note its a case-insensitive substring search.

Also, specifying behaviour around character normalization (e.g. `U` matching `Ü`) would be helpful.

Author: I mean, the example is pretty clear I think? Character normalization is a good point though, we probably _should_ do this.

 - Ability to filter by space.
 - Ability to filter by room name.

These goals shaped the design of this proposal. 

## Proposal

At a high level, the proposal introduces a way for clients to filter and sort the rooms they are
joined to and then request a subset of the resulting list of rooms rather than the entire room list.
```
         All joined rooms on user's account
Q W E R T Y U I O P L K J H G F D S A Z X C V B N M
\                                                 /
 \                                               /
  \      Subset of rooms matched by filters     /
   Q W E R T Y U I O P L K J H G F D S A Z X C V
                       |
   A C D E F G H I J K L O P Q R S T U V W X Y Z     Rooms sorted by name (or by recency, etc)
   |_______|
       |

   A C D E F                                         first 5 rooms requested
```
It also introduces a number of new concepts which are explained in more detail later on:
 - Core API: The minimal API to be sync v3 compatible.
 - Extensions: Additional APIs which expose more data from the server e.g presence, device messages
   or additional sort/filter operations.

## Core
A complete sync request looks like:

> `POST /v3/sync?pos=4`:

Reviewer 2: where has the timeout param gone?

Author: It's there in the proxy, will add.

```js
{
  // Sliding Window API
  "lists": [
    {

>       "rooms": [ [0,99] ],

Reviewer 2: It feels weird that this describes a set of ranges, but is called `rooms` (which would imply it contained room IDs or room objects of some kind).  Shouldn't it be called `ranges`?

>       "rooms": [ [0,99] ],

Author: It's slightly more bandwidth efficient to use an array whilst still preserving readability in that the start index will always be before the end index. If we did it the way you suggest, it's entirely valid to see `"rooms": [ {"end": 99, "start": 0} ]` which just becomes confusing.

>       "rooms": [ [0,99] ],

Reviewer 3: Also regarding bandwidth efficiency, this is sticky no? You won't have to send this up to the server very often anyway.

>       "rooms": [ [0,99] ],

Author: This field is not sticky and changes whenever you scroll the room list.

> CBOR or protobuf should be used instead an call it a day.

This isn't a valid suggestion. Protobuffers aren't suitable for extensible keys that Matrix support. CBOR encodes key names as strings unless you enumify them like MSC3079 does (which I also wrote). The difference overall is small however.

>       "sort": [ "by_notification_count", "by_recency", "by_name" ],

Author: I don't think they will all necessarily have the `by_` prefix e.g `alphabetically_by_name`. I'd like to see different sort extension MSCs get proposed first before we assume the `by_` prefix is sensible.

>       "sort": [ "by_notification_count", "by_recency", "by_name" ],

Author: As the MSC states, it isn't possible to do ASC/DESC on the sort orders. It's something we could add if there is a need for it though.

>       "sort": [ "by_notification_count", "by_recency", "by_name" ],

Reviewer 3: If this land, there would be no nice way to do so. Best to just leave it encoded into the string as it is now and explicitly state it as such.

      "required_state": [

>         ["m.room.join_rules", ""],

Reviewer: It's technically a tuple of 2, but JSON doesn't know about tuples.

>         ["m.room.join_rules", ""],

Reviewer 3: Sure but I think using a JSON object is more correct.

>         ["m.room.join_rules", ""],

Author: The same reasons that ranges are arrays and not objects. It's generally more common to see the event type prior to the state key hence why the event type is first, but as @ShadowJonathan points out, they are technically tuples.

        ["m.room.history_visibility", ""],
        ["m.space.child", "*"]
      ],
      "timeline_limit": 10,
      "filters": {
        "is_dm": true
      }
    }
  ],

  // Room Subscriptions API

>   "room_subscriptions": {

Reviewer 2: shouldn't this be `subscribe_rooms` for symmetry with `unsubscribe_rooms`, and to make it clear that it's an action, rather than being a sticky list?

Reviewer 2: it's also weird that this is a hash of id->objects, but unsubscribe is a list of ids. might be cleaner to have them both be a list of objects?

Reviewer: > ...might be cleaner to have them both be a list of objects?

I'm not sure, `room_subscriptions`/`subscribe_rooms` semantically relates to `PUT` for me, while `unsubscribe_rooms` relates to `DELETE`, where convention is with `PUT` to push all data you'd want to "store", and with `DELETE` you only reference where you've put the data, and that you want to delete it.

      "!sub1:bar": {
          "required_state": [ ["*","*"] ],
          "timeline_limit": 50
      }
  },
  "unsubscribe_rooms": [ "!sub3:bar" ]

  // Extensions API
  "extensions": {}
}
```
An entire response looks like:
`HTTP 200 OK`
```js
{
  // Connection and Streaming API

>   "initial": true,

Reviewer 2: the `initial` field is undocumented, as far as i can see?

>   "pos": "5",

Reviewer 4: `pos` is a number in the proxy - should it be treated as a string?

Reviewer 2: sounds like a proxy bug. the spec here is quite unambiguous that it should be an opaque string.

  // Sliding Window API
  "ops": [
    {
      "list": 0,
      "range": [0,99],
      "op": "SYNC",
      "rooms": [
        {
          "room_id": "!foo:bar",

>           "name": "The calculated room name",

Author: The server isn't forcing the room name on the client in that the client is free to use whatever name it wants for a given room. The server is however forcing a sort order on the rooms based on the canonicalised room name. Sync v3 doesn't return heroes/member counts/etc because the calculation is done server-side (it needs to for the sort order) so we just conveniently return the calculated room name in this `name` field.

            {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!foo:example.com", "content":{"via":["example.com"]}},
            {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!bar:example.com", "content":{"via":["example.com"]}},
            {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!baz:example.com", "content":{"via":["example.com"]}}
          ],
          "timeline": [
            {"sender":"@alice:example.com","type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"invite"}},
            {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
            {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
            {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"C"}},
            {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"D"}},
          ],
          "notification_count": 54,
          "highlight_count": 3
        },
        // ... 99 more items
      ],
    }
  ],

>   "counts": [1337], 

Reviewer 2: I wonder if 'counts' is the best term here - isn't it more a `length`?

Reviewer 4: or `total` to preserve the meaning of the number.

Reviewer: Related: we should probably have a document somewhere in the spec repo that standardises these for these kinds of ambiguous identifiers, i.e. "between `len`, `length`, `total`, `num`, `count`, etc., which one should the matrix spec use?"

Reviewer 2: trying to implement this, i'm unclear on what the total actually is of - doesn't it make more sense if it's scoped to a given `list`? currently it's a global, which implies that it's just telling you the total number of rooms that are on the user's radar, which doesn't seem like a very useful thing to know.

Reviewer 4: it's the number of rooms which could be in the list. In the bot-sdk's test impl I use it to figure out how many ranges (windows) to ask for: https://github.com/turt2live/matrix-bot-sdk/blob/b04dcecfa99c183f51d3777ff00d0fc6b47ab4d4/src/syncv3/V3List.ts#L100

(though it's questionable that the bot-sdk asks for ranges - haven't quite figured out if there's performance/network gains to be had here)

Author: Having the total room count is useful for clients to know how many rooms there are for figuring out how to size list views. It's also useful for bots/other clients if they just want to grab all rooms.

The total room count exists at the top level because there is no list scope - only ops (and each op is then scoped to a list).

Happy to rename this to `totals`.

Reviewer 2: oh, so you get a total count of rooms per list, one item in the array per list? i think i missed that the value is an array. (and eitherway, its behaviour needs to be spelt out in the MSC)

  // Room Subscriptions API
  "room_subscriptions": {

>     "!sub1:bar": {

Reviewer 2: It feels a bit clunky and inconsistent that the room subs returns rooms as key-values keyed by room ID, but the sliding window list has them as a list of room objects.  If they were the same, then the implementation could use the same datastructure for both, i think?

Author: Not sure I follow you. We need to use a list of room objects for the sliding window as ordering is important. Ordering is irrelevant for room subscriptions, so whilst yes we could put this as an array of room objects it doesn't really help beyond being able to reuse a very small portion of the response struct, at the cost of having to answer questions like:
 - is the ordering in `room_subscriptions` important?
 - is it deterministic? If so, what's the sort order?
 - what do you do if the same room appears in the list twice?
 - why do I have to loop through the entire list containing all my subscriptions just to find a very specific room I asked for?

These questions disappear if it's a key-value map.

Reviewer 5: Maybe both can be key value maps if we separate list order metadata from content, see https://github.com/matrix-org/matrix-doc/pull/3575#discussion_r777030851

Reviewer 6: ^ This seems to be actually pretty good idea.

Why not just have one merged cumulative object of rooms for every "query type" and then have ids in each list/map.

        {"sender":"@alice:example.com","type":"m.room.member", "state_key":"@alice:example.com", "content":{"membership":"join"}}
        ],
        "timeline": [
        {"sender":"@alice:example.com","type":"m.room.create", "state_key":"", "content":{"creator":"@alice:example.com"}},
        {"sender":"@alice:example.com","type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"invite"}},
        {"sender":"@alice:example.com","type":"m.room.history_visibility", "state_key":"", "content":{"history_visibility":"joined"}},
        {"sender":"@alice:example.com","type":"m.room.member", "state_key":"@alice:example.com", "content":{"membership":"join"}}
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
        ],
        "notification_count": 1,
        "highlight_count": 0
    }
  },

  // Extensions API
  "extensions": {}
}
```
These fields and their interactions are explained in the next few sections. This forms the core of
the API. Additional data can be returned via "extensions". 

### Connections and streaming data

> At a high level, the syncing mechanism creates a "connection" to the server to allow the

Author: This should really be treated as a connection though in that if WebScokets support is sprouted from this MSC (the design allows it) then the "connection" would actually be a real WS connection. It only isn't at present because Matrix uses long-polling.

> At a high level, the syncing mechanism creates a "connection" to the server to allow the

Reviewer 3: Well the WebSockets MSC doesn't exist yet and may look very different from what you envision.
I don't think making decisions based on a design that no one has seen or reviewed is a good idea. It makes the future rigid and the present awkward.

That aside, the websocket MSC can say something like, a sync session is only valid for the duration of the websocket connection. At least then, when I speak to people I can say "sync session" and abstract away long polling vs websocket vs future alternatives.

Basically what I'm saying is, a Matrix specific name for this connection concept won't hurt.

the client is currently at. Unlike `/sync` v2, this token is ephemeral and can be invalidated at any
time. When a client first connects to the server, no `pos` is specified. Also unlike `/sync` v2, this
token cannot be used with other APIs such as `/messages` or `/keys/changes`. Note that the "connection"
formed to the server is _not_ a long-lived TCP connection, it is just an application-level concept
of a connection.

In simple servers, the `pos` may be an incrementing integer, but more complex servers may use vector
clocks or contain node identifying information in the token. Clients MUST treat `pos` as an opaque
value and not introspect it.

When a `pos` is invalidated and the client attempts to use the `pos`, the server MUST send back a
standard error response as a HTTP 400 containing:
```js
{
  "error": "Unknown position",
  "errcode": "M_UNKNOWN_POS"
}
```
This then allows the client to reset their connection and send an initial request (with all sticky
request parameters) without a `pos` value to restart the connection. The flag `initial: true` MUST
be set in the response when there is no `pos` value provided. When there is a valid `pos`, this flag

> MUST be omitted (sending `initial: false` is wasteful).

Reviewer 5: MUST is extreme for no reason. Isn't SHOULD okay?

Author: It's easy for server developers to do this, it's just often forgotten about. A compliance test suite can easily test this, so I feel MUST is acceptable given how noisy/bandwidth consuming sending `initial: false` over and over would be. It's also partly because it's horrible seeing an "empty" sync v2 response contain tons of fields just like this for no real reason.

Reviewer 2: +1 to mandating that servers shouldn't include redundant data in sync responses, to avoid all the problems we've had with v2 sync responses containing loads of blank redundant data which then break clients when removed.

Reviewer 3: What should clients do then -- reject the sync response if the redundant field is sent? In reality, if a server does return this (intentionally or not) clients will simply let it work anyway. Using MUST here would be a guideline disguised as important spec.

> A response for a given `pos` must be idempotent to account for packet loss. For example:

Reviewer 5: Why? If the client tried again that means it did not receive the previous update. Then it should be okay to send a 'better' update instead (e.g. if a user stopped typing, we can leave out that EDU)

Author: This is a good point, in that if the client doesn't update `pos` then it should be safe to assume they didn't get the original response. Come to think of it, sync v2 does exactly this in that syncing with a `since` token can produce different responses depending on when you sync. I'll remove these bits then.

               | <-------HTTP 200 OK---------- |
```
For this reason, some request parameters are not suitable to be made "sticky". These include parameters
which are extremely dynamic in nature, such as list ranges.

### Room List parameters

One or more room lists can be requested in sync v3 like so:
```js
{
  "lists": [
    {
      // Sliding window ranges, see the Sliding Window API for more information.
      "rooms": [ [0,99] ],
      // Sticky. List sort order. See Sliding Window API for more information.
      // These fields may be expanded through use of extensions.
      "sort": [ "by_notification_count", "by_recency", "by_name" ],

      // Sticky. Required state for each room returned. An array of event type and state key tuples.
      // Note that elements of this array are NOT sticky so they must be specified in full when they
      // are changed.
      "required_state": [
        // Request the join rules event. Note that the empty string is required here to match
        // the event's blank state_key.
        ["m.room.join_rules", ""],
        ["m.room.history_visibility", ""],
        // Request all `m.room.member` state events.

>         // The * is a special sentinel value meaning 'all keys'.

Author: Skipping the second element makes it harder for statically typed languages to represent the API shape accurately. For example, in the proxy (in Go) this field is literally a list of 2-element string arrays:
```go
RequiredState [][2]string     `json:"required_state"`
```
This becomes impossible to represent if we allow 1 _or_ 2 values, pushing more validation work onto the developer.

>         // The * is a special sentinel value meaning 'all keys'.

Reviewer 3: It's good you bring up statically typed languages, so I know it's safe to use as an argument now.

This concept of a list of 2-element string arrays doesn't exist in many popular languages. Or tuples for that matter.
This extra validation work you describe, I am going to have to do it for Kotlin, since Kotlin doesn't have tuples or statically sized lists. It also cannot automatically parse an array into a class, since it well formed JSON shouldn't look like that, I'm going to have to do extra work for it.

This should really just be an object (and not an array), as every statically typed language will understand that.

        // Note that `*` is NOT a generic glob function. You cannot specify `foo*` to pull in keys
        // like `food` and `foobar`. In this case, the * is treated as a literal *.

>         ["m.room.member", "*"]

Reviewer 4: as a lazy shortcut, it might be useful to have a character/symbol to denote "current user", or just have the API always assume the client wants to know the user's current membership. Eg: add `["m.room.member", "$current"]` by default

Reviewer: I think that has a potential to clash, why not a different value type? iirc anything other than a `string` is illegal as a state_key, so this shouldn't conflict.

Reviewer 4: Not all architectures allow for such an easy check. In my test implementation, it was so difficult that a symbol was used anyways: https://github.com/turt2live/matrix-bot-sdk/pull/189

Adding a second symbol to the spec is probably the wrong answer, but assuming the client wants to know about the membership event *always* isn't too much of a stretch I think.

Author: What's the use case for wanting the m.room.member for the current user? I would have expected people to typically split `is_invite: true` and `is_invite: false` into different lists so you would know the membership based on context (which list the room is present in)?

>         ["m.room.member", "*"]

Reviewer 3: Another literal here could be something like `"LAZY"`, which means, Only give me state events if the `state_key` matches the userId of any events returned in the room timeline.

Reviewer 2: i quite like this as a way to indicate LL member semantics (more so than having it as a filter), but you must remember that the semantics are more than just "only give me members for events in the room timeline" - it also has to cover the behaviour during gappy syncs.

Author: What behaviour are you alluding to exactly @ara4n ?

Reviewer 4: From the [`/sync` endpoint](https://spec.matrix.org/v1.1/client-server-api/#get_matrixclientv3sync):

> When a sync is `limited`, the server MUST return membership events for events in the gap (between `since` and the start of the returned timeline), regardless as to whether or not they are redundant. This ensures that joins/leaves and profile changes which occur during the gap are not lost.

        // If unset, both DM rooms and non-DM rooms are returned. If false, only non-DM rooms
        // are returned. If true, only DM rooms are returned.
        "is_dm": true,
        // A list of spaces which target rooms must be a part of. For every invited/joined room for
        // this user, ensure that there is a parent space event which is in this list. If unset, all
        // rooms are included. Servers MUST NOT navigate subspaces. It is up to the client to
        // give a complete list of spaces to navigate. Only rooms directly in these spaces will be
        // returned.
        "spaces": ["!foo:bar", "!bar:baz"],
        // Flag which only returns rooms which have an `m.room.encryption` state event. If unset,
        // both encrypted and unencrypted rooms are returned. If false, only unencrypted rooms
        // are returned. If true, only encrypted rooms are returned.
        "is_encrypted": true,
        // Flag which only returns rooms the user is currently invited to. If unset, both invited
        // and joined rooms are returned. If false, no invited rooms are returned. If true, only
        // invited rooms are returned.
        "is_invite": true

>       }

Reviewer 2: Are we missing a filter for showing historical rooms?

Reviewer 2: ...and showing knocked rooms?

Author: We're missing a lot of filters, but these can likely be punted to extension MSCs.

Reviewer 2: okay, but i think you should spell out the ones which are being punted, to stop everyone reviewing this MSC tripping over the same question.

    {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!foo:example.com", "content":{"via":["example.com"]}},
    {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!bar:example.com", "content":{"via":["example.com"]}},
    {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!baz:example.com", "content":{"via":["example.com"]}}
  ],
  // Last event is most recent. Max timeline_limit events.
  "timeline": [
    {"sender":"@alice:example.com","type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"invite"}},
    {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
    {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
    {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"C"}},
    {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"D"}},
  ],
  "notification_count": 54, // same as sync v2
  "highlight_count": 3      // same as sync v2
}
```

_Rationale: The room name and counts are required for display on the UI. They are calculated server

> side because they are required for sort operations on lists. The `required_state` is controversially

Reviewer 2: "required" state, past tense, makes it sound like this is the state required to process the response - i.e. the state at the beginning of the response, not the end.  What is the word 'required' up to here?  Wouldn't it be better to call it 'current state' or something?

Reviewer: I'm proposing `state_view`/`state_subset`, as essentially sync is only returning a subset of state of the room, and this parameter is restricting it to that.

"Required" state threw me off the first time as well, but only because it qualified as a filter in my head, as then the room would be "required" to have X state before its synced, but that wasn't so.

Author: `required_state` in the response matches the same field option in the request where "required state" means the state required by the client. I'm happy to rename this to something clearer, especially given @ShadowJonathan remarks about "the room would be "required" to have X state before its synced" which is a good point.

the **current state** which breaks from sync v2 which has the `state` be "the state before the start
of the timeline". Sync v2's rationale was event duplication (state events can appear in both the
state section and the timeline section if it's the current state) and the fact that clients would have
to rewind state to work out historical display names. Clients who show historical display names
already need to rewind state by inspecting the `prev_content` of an event to display text like
"@alice changed their name from Alice to Alice2". Event duplication may be
reduced using Event ID -> Event maps in the response should this be a concern. The benefit of

> returning the current state is that servers can cache the latest state to return the response more

Author: You're right that idempotency means servers need to remember responses, but this is critically just the _last_ response, not the response for every single `pos` ever sent to the client. Idempotency guarantess are there purely to account for packet loss.

> returning the current state is that servers can cache the latest state to return the response more

Reviewer 3: It looks like you're about to drop the idempotent requirement anyway, so my comment here won't matter anyway.

quickly. If, instead, servers returned the state at the start of a timeline block, servers are forced
to either rewind this state (as clients will need to do) or worse, do an expensive database access to
request the state before an event. As clients can be at different points in the stream for a given
room, this would force servers to cache every possible room state. It's not practical for servers to
cache every single possible earlier state for each room._

### Sliding Window API

At a high level, the sliding window API provides a way to synchronise a subslice of a list in a
bandwidth efficient way. It does this by referring to "operations" which must be performed on the
stored client list, such as INSERT, DELETE and UPDATE. Each operation has an index position OR a
range of index positions which tells the client where the operation should be performed. The

> possible operations are:

Reviewer 5: I think the set of actions can be improved if we require that the list has a fixed maximum size and empty slots must be at the end.

Reviewer 7: That doesn't work for alphabetic ordering, I would say.

 - `SYNC`: Sets a *range* of entries. Clients SHOULD discard what they previous knew about entries in
   this range.
 - `INSERT`: Sets a *single* entry. If the position is not empty then clients MUST move entries
   to the left or the right depending on where the closest empty space is.

>  - `DELETE`: Remove a *single* entry. Often comes before an `INSERT` to allow entries to swap places.

Reviewer 5: How about a MOVE that does DELETE+INSERT in one operation? Also 'swap' is the wrong word here, because it will move all entries inbetween

Author: It's not really a move operation though, in that `DELETE 4, INSERT 0 != MOVE 4,0`. You really are deleting an entry and inserting a new entry which you previously had no knowledge about. We could absolutely mux these into an additional operation, but does that really help things? You still need a `DELETE` for actual deletions (leaving rooms) and `INSERT` for actual insertions (joining new rooms) so why not just reuse the core machinery we already need? This reminds me a bit of CISC vs RISC debate.

 - `UPDATE`: Update a *single* entry. Updates are cumulative (consisting of deltas only).
 - `INVALIDATE`: Remove a *range* of entries. Clients MAY persist the invalidated range for offline
   support, but they should be treated as empty when additional operations which concern indexes in
   the range arrive from the server.

For example:
```
            Client                         Server
          []  |                              |  0,1,2,3,4,5,6,7,8   index
              |                              | [A,B,C,D,E,F,G,H,I]
              | -------- range[0,4] -------> |
 [A,B,C,D,E]  | <--- SYNC[0,4]=A,B,C,D,E --- |
              |                              |  0,1,2,3,4,5,6,7,8
              |                              | [H,A,B,C,D,E,F,G,I]  H moves to the front
              | ----- wait for updates ----> |

>  [H,A,B,C,D]  | <- DELETE[4], INSERT[0]=H--- |

Reviewer 5: What happens if you don't delete? If we say the list has a fixed size, then the client can figure out it must drop off the end.

Reviewer 5: Hmm do you mean this?: The client sends a sliding window update. Before that is handled by the server, the server already sends a sync. The client assumes the sync is received already uses the updated sliding window, but this is wrong.

              |                              | [H',A,B,C,D,E,F,G,I] H is updated to H'
              | ----- wait for updates ----> |
[H',A,B,C,D]  | <------ UPDATE[0]=H'-------- |
              |                              |  0,1,2,3,4,5,6,7,8
              |                              | [J,K,L,M,N,O,P,Q,R]  Entire list is replaced
              | ----- wait for updates ----> |
 [J,K,L,M,N]  | <----INVALIDATE[0,4]-------- |

>               |      SYNC[0,4]=J,K,L,M,N     |

Author: Yes and no. It's useful to have ways to delete ranges of rooms (say if the list filters change then we want to invalidate large chunks of the list) which is why INVALIDATE exists in the first place. We could implicitly say "if you get a SYNC for a range which you already think you have data for, replace that data", so that way we can remove the INVALIDATE call. I would need to think through the implications of making this implicit. This might be okay so long as we never have cases where invalidate ranges != sync ranges.

              |                              | [J,K,L,N,O,P,Q,R]    M is deleted
              | ----- wait for updates ----> |
 [J,K,L,N,O]  | <- DELETE[3], INSERT[4]=O--- |
```
The sync v3 API exposes this API shape via the following request parameters:
```js
{
  // Multiple lists can be requested
  "lists": [
    {
      // Multiple sliding windows inside a list can be requested. Integers are _inclusive_.

>       "rooms": [ [0,9], [20,29] ],

Reviewer 5: The only reason multiple windows are allowed is for some clients that display multiple windows at once. But those could just use multiple "lists".
This means the "rooms" field can describe just a single window (or it can be split into two fields like "start" and "amount").

Reviewer 5: This might lead to duplicate data for when the windows overlap. To fix this, we can separate list metadata and actual content. See https://github.com/matrix-org/matrix-doc/pull/3575/files#r777030851

Author: We could do this and in fact early designs did have this. However, those multiple "lists" _are intrinsically coupled_, resulting in synchronised churn on the list metadata when rooms change positions, as well as having to specify the request params (required state, timeline limit, etc) N times for each tracked range. It just made more sense to allow multiple windows.

For usage context, the majority of clients will use multiple sliding windows (at present at least): one window tracking the top N rooms and another tracking the scroll position of the list UI. Clients currently have to track the top N rooms to get notifications (assuming a most recent sort order) as the `notifications` API isn't fully fleshed out yet.

I'm open to ideas to allow lists to compose together better than at present, if it provides a more simplistic API shape whilst still being easy to follow.

      // How the list should be sorted on the server. The first value is applied first, then tiebreaks
      // are performed with the 2nd sort order, then the 3rd until there are no more sort orders left.
      "sort": [ "by_notification_count", "by_recency", "by_name" ],
      // Additional Room List request parameters omitted as they are
      // unrelated to the semantics of the sliding window, see previous section.
    }
  ],
}
```
Which returns the following response parameters:
```js
{
  "ops": [
    {
      // Which list is affected by this operation
      "list": 0,
      // Which index positions are affected. These are both inclusive.
      "range": [0,9],
      // The operation being performed on these index positions.
      "op": "SYNC",
      // The data to put in these positions.

>       "rooms": [

Reviewer 5: Why are the list ordering and the room updates coupled so much? This will lead to duplication if multiple lists contain the same rooms. Can we separate ordering and actual updates?

Author: We already do separate things a bit in that if a room is present in the subscription and the list, only the subscription gets the update and the list just gets the room ID in question. We could go the whole hog here and put all room updates in a map and just keep list ordering bare bones, but I don't see the value here.

> This will lead to duplication if multiple lists contain the same rooms.

This is by design as the lists may have different `required_state` and `timeline_limit`. Event duplication is a concern which can be alleviated via event maps of event_id -> Event JSON, but I'm wary about doing that at this point whilst the API is still in flux, for precisely these sorts of reasons (maybe there's better ways to de-dupe naturally in the API).

Reviewer 5: > This is by design as the lists may have different required_state and timeline_limit

It's easy to merge the requirements: union over required_states, max over timeline_limits (although "limit" is not the right word then, maybe timeline_limit can be renamed to timeline_request?)

Author: We now do this. :)

        },
        // ... 9 more items
      ],
    }
  ],
  // The total number of entries in the list. Index positions match up to the lists in the request.
  "counts": [1337],
}
```

The possible `sort` operations are:
 - `by_recency`: Sort by `origin_server_ts` on the most recently _received_ event in the room. Note
   that due to clock drift over federation it is possible for rooms to re-order such that the most
   recently received event in the entire list does not cause that room to go to index position 0.
 - `by_highlight_count`: Sort by the `highlight_count` for this user in this room, which is the
   number of unread notifications for this room with the highlight flag set. This value is also present
   in sync v2.
 - `by_notification_count`: Sort by the `notification_count` for this user in this room, which is the
   total number of unread notifications for this room. This value is also present in sync v2.
 - `by_name`: Sort by room name lexicographically. This requires servers to implement the
   [room name calculation algorithm](https://matrix.org/docs/spec/client_server/latest#calculating-the-display-name-for-a-room).

>    The server MUST perform the following steps:

Reviewer 3: This somewhat answers my earlier question. I see below that clients that have more bespoke calculations will just not use this.

     * Calculate the room name from this user's perspective. This may vary depending on the user as
       DM rooms will have the room name set to the name of the _other user_. This is the value that
       will be returned in the `name` field for the room but is NOT the value that the server should
       perform sort operations on. See following steps.
     * Remove any of the following characters from the beginning/end of the calculated name: `#!()):_@`.
       This ensures things like canonical aliases display in roughly the right alphabetical locations
       rather than all together with all rooms that start with `#`.
     * Lower-case the result by unicode. This ensures `Matrix` and `matrix` sort in the same locations.
     * Perform sort operations on this 'canonicalised' name.

Sorting algorithms MUST be stable and deterministic to avoid needless churn as otherwise identical
rooms keep swapping positions. This can easily be achieved by including a final tiebreak based on the
room ID (e.g lexicographical sort on the room ID) to guarantee stability and determinism. It is currently

> _not possible_ to invert the sort order (ASC vs DESC). This may be added to this MSC if there is a

Reviewer 3: Ah this somewhat answers my earlier question. I still don't know what the existing order is though; for the inversion that is.

For the first of these issues, the sync v3 API exposes a "room subscription" API. For the second issue,
the sync v3 API exposes a "notifications" API.

### Room Subscription API

Sometimes clients know exactly which room they want to get information about e.g by following a
permalink or by refreshing a webapp currently viewing a specific room. The sliding window API alone
is insufficient for this use case because there's no way to say "please track this room explicitly".
The room subscription API serves as a way to provide this tracking. At a high level, the client
provides a map of room ID to room list parameters and the server then returns the response in the
same format as the sliding window API, just without the operations/indexes.

To track a room `!sub1:bar`, the client would send the following request:
```js
{
  "room_subscriptions": { // sticky
      "!sub1:bar": { // sticky
          "required_state": [ ["*","*"] ],
          "timeline_limit": 50
      }
  }
}
```
This would return the following response:
```js
{
  "room_subscriptions": {
    "!sub1:bar": {
        "name": "Alice and Bob",
        "required_state": [

>         {"sender":"@alice:example.com","type":"m.room.create", "state_key":"", "content":{"creator":"@alice:example.com"}},

Reviewer 2: Is the indenting deliberately wrong/inconsistent on these? i find myself getting distracted - should we just stick to 2-space-indents in the examples?

        {"sender":"@alice:example.com","type":"m.room.member", "state_key":"@alice:example.com", "content":{"membership":"join","displayname":"Alice"}},
        {"sender":"@alice:example.com","type":"m.room.member", "state_key":"@bob:example.com", "content":{"membership":"join","displayname":"Bob"}}
        ],
        "timeline": [
        {"sender":"@alice:example.com","type":"m.room.create", "state_key":"", "content":{"creator":"@alice:example.com"}},
        {"sender":"@alice:example.com","type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"invite"}},
        {"sender":"@alice:example.com","type":"m.room.history_visibility", "state_key":"", "content":{"history_visibility":"joined"}},
        {"sender":"@alice:example.com","type":"m.room.member", "state_key":"@alice:example.com", "content":{"membership":"join","displayname":"Alice"}},
        {"sender":"@alice:example.com","type":"m.room.member", "state_key":"@bob:example.com", "content":{"membership":"join","displayname":"Bob"}}
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},

>         ],

Reviewer 2: Do we not need the concept of a limited sync still still?

Hydrogen's grid view. The `unsubscribe_rooms` array allows rooms to be efficiently deleted from the
map. An alternative would be to specify an empty JSON object in the room subscription but that feels
less explicit than the array form._

### Notifications API

If you are tracking the top 5 rooms and an event arrives in the 6th room, you will be notified about
the event ONLY IF the sort order means the room bumps into the top 5. If for example you sorted
`by_name` then you won't be notified about the event in the 6th room, unless it's an `m.room.name`
event which moves the room into the top 5. In most "recent" sort orders a new event *will result* in
the 6th room bumping to the top of the list. A notable exception is when the rooms are sorted in
*alphabetical order* (`by_name`), which is what some other chat clients do for example. In this case,
you don't care about the event unless the event is a "highlightable" event (e.g direct @mention).
The notifications API exists to provide a mechanism for clients to display "unread messages"
indicators on the room list at positions not currently inside a sliding window.

TODO:

> - Unsure how much data to expose (probably index position + notif/highlight counts?). If we do _counts_

Reviewer 5: Is the idea that we only send a little data and if the client wants to know more they can subscribe to the room?

Author: Exactly.

then we are doomed to send a response to a client every time an event is sent in a noisy room, which

> seems rather wasteful. Perhaps make it configurable?

Reviewer 5: Maybe we send an approximate count, like only two digits of precision (21 -> 21, but 1234 -> 1200)

- This has not been fully specced yet because in practice most clients sort by recency so it's not
  urgent to include this. For clients who sort by name though, this is a show stopper.

> ### Bandwidth optimisations for persistent clients

Reviewer 2: I think we should clarify how this interacts with lazy-loading members (#1227). The current wording seems to imply that required_state will include all members, which could be super wasteful and kill all the benefits of LL members?

Author: Correct, there is no special lazy room member handling in this MSC yet. I suspect I would add this in as an extension as it's a custom filter effectively.

Reviewer 2: okay, it could be factored out to aid review, but we need to flag loud and clear that this deliberately doesn't implement LL members yet - as in that respect it's very much a regression.

 - Ability to filter by space.
 - Ability to filter by room name.

These goals shaped the design of this proposal. 

## Proposal

At a high level, the proposal introduces a way for clients to filter and sort the rooms they are
joined to and then request a subset of the resulting list of rooms rather than the entire room list.
```
         All joined rooms on user's account
Q W E R T Y U I O P L K J H G F D S A Z X C V B N M
\                                                 /
 \                                               /
  \      Subset of rooms matched by filters     /
   Q W E R T Y U I O P L K J H G F D S A Z X C V
                       |
   A C D E F G H I J K L O P Q R S T U V W X Y Z     Rooms sorted by name (or by recency, etc)
   |_______|
       |

   A C D E F                                         first 5 rooms requested
```
It also introduces a number of new concepts which are explained in more detail later on:
 - Core API: The minimal API to be sliding sync compatible.
 - Extensions: Additional APIs which expose more data from the server e.g presence, device messages
   or additional sort/filter operations.

## Core
A complete sync request looks like:

> `POST /_matrix/client/unstable/org.matrix.msc3575/sync?pos=4&timeout=30000`:

Reviewer 5: We should use this opportunity and remove the query parameters and move them into the json request so that clients don't have to send an OPTIONS request every time.

Reviewer 5: See https://github.com/matrix-org/synapse/issues/9669

Author: Right, so this is only when the server sends back `Access-Control-Max-Age` header.

**I love it.** Will implement this now in the proxy.

Author: Hmm, I'm not seeing OPTIONS requests with every /sync in the proxy. These requests are POST requests. Will need to investigate this more before moving the pos to the HTTP body.

```js
{
  // Sliding Window API
  "lists": [
    {
      "ranges": [ [0,99] ],
      "sort": [ "by_notification_count", "by_recency", "by_name" ],
      "required_state": [
        ["m.room.join_rules", ""],
        ["m.room.history_visibility", ""],
        ["m.space.child", "*"]
      ],
      "timeline_limit": 10,
      "filters": {
        "is_dm": true
      }
    }
  ],

  // Room Subscriptions API
  "room_subscriptions": {
      "!sub1:bar": {
          "required_state": [ ["*","*"] ],
          "timeline_limit": 50
      }
  },
  "unsubscribe_rooms": [ "!sub3:bar" ]

  // Extensions API
  "extensions": {}
}
```
An entire response looks like:
`HTTP 200 OK`
```js
{
  // Connection and Streaming API
  "initial": true,
  "pos": "5",

  // Sliding Window API
  "ops": [
    {
      "list": 0,
      "range": [0,99],
      "op": "SYNC",
      "rooms": [
        {
          "room_id": "!foo:bar",
          "name": "The calculated room name",

>           "required_state": [

Reviewer 5: Currently there is no way to tell clients about removal of state events. How about a second field with "dropped_state" that lists all the state event ids that have been dropped and not replaced since the last update

Author: I'd rather we expressed removal of state events as either:
 - The previous state event is returned, or
 - An empty `content: {}` is sent if the state is deleted entirely.

I prefer this because it means clients will handle deletions automatically as soon as they can handle setting state, without having to add in extra code.

Reviewer 8: Setting `content: {}` is more of a redaction than a deletion, which is a subtly different thing. I'm not 100% sure that we absolutely need the distinction, but I'm wary of designing out the ability to express it.

I don't know whether these sort of changes would be shown in the timeline UI at all.

 [J,K,L,M,N]  | <----INVALIDATE[0,4]-------- |
              |      SYNC[0,4]=J,K,L,M,N     |
              |                              | [J,K,L,N,O,P,Q,R]    M is deleted
              | ----- wait for updates ----> |
 [J,K,L,N,O]  | <- DELETE[3], INSERT[4]=O--- |
```
The sync v3 API exposes this API shape via the following request parameters:
```js
{
  // Multiple lists can be requested
  "lists": [
    {
      // Multiple sliding windows inside a list can be requested. Integers are _inclusive_.
      "ranges": [ [0,9], [20,29] ],
      // How the list should be sorted on the server. The first value is applied first, then tiebreaks
      // are performed with the 2nd sort order, then the 3rd until there are no more sort orders left.
      "sort": [ "by_notification_count", "by_recency", "by_name" ],
      // Additional Room List request parameters omitted as they are
      // unrelated to the semantics of the sliding window, see previous section.
    }
  ],
}
```
Which returns the following response parameters:
```js
{
  "ops": [
    {
      // Which list is affected by this operation
      "list": 0,
      // Which index positions are affected. These are both inclusive.

>       "range": [0,9],

Reviewer 5: Are these indexes of the sliding window or of the entire list?

Author: Index positions and ranges always refer to the entire list.

      // The operation being performed on these index positions.
      "op": "SYNC",
      // The data to put in these positions.
      "rooms": [
        {
          "room_id": "!foo:bar",
          "name": "The calculated room name",
          // Additional response parameters omitted as they are
          // unrelated to the semantics of the sliding window.
          // See previous section on room list parameters.
        },
        // ... 9 more items
      ],
    }
  ],
  // The total number of entries in the list. Index positions match up to the lists in the request.
  "counts": [1337],
}
```

The possible `sort` operations are:
 - `by_recency`: Sort by `origin_server_ts` on the most recently _received_ event in the room. Note
   that due to clock drift over federation it is possible for rooms to re-order such that the most
   recently received event in the entire list does not cause that room to go to index position 0.

>    The highest `origin_server_ts` value comes first in the list.

Reviewer 7: Does this always match what the client is sent as the most recent event in the room? In the traditional sync, a clean, new sync would reorder recently received events to the back, when their ts is older. Would that affect this sorting or would it always stick to he most recently received event on the server side?

 - `by_highlight_count`: Sort by the `highlight_count` for this user in this room, which is the
   number of unread notifications for this room with the highlight flag set. This value is also present
   in sync v2. The highest `highlight_count` comes first in the list.
 - `by_notification_count`: Sort by the `notification_count` for this user in this room, which is the
   total number of unread notifications for this room. This value is also present in sync v2.

>   The highest `notification_count` comes first in the list.

Reviewer 7: Nheko sorts by `["by_has_highlights", "by_notification_count",  "by_recent_message", "by_room_id"]`

Sorting by highlight count is very confusing, because there is not really a way for users to understand that sorting. They only see a red or blue bubble and that bubble always has the notification count, not the highlight count.

And sorting by most recently received event is annoying, because it makes membership changes bump the room list (and other state events you don't care about). As such Nheko only uses the timestamp of events, that it actually shows in the room list.

Lastly it uses the room id as the last effort tie breaker. That is so unlikely to happen, that it probably does not need to be spelled out in the sync sorting, I just included it for completeness sake. I'm guessing many servers will do that anyway.

Author: Extension MSCs can include these custom sort orders.

connection, or re-requesting a page that was previously INVALIDATEd, the server will perform the
following operations:
 - For this device: check the last sent event ID for the room ID in question. Count the number of
   timeline events from that point to the latest event. Call it `N`.
 - For this specific sync request: calculate a reasonable upper-bound for how many events will be
   returned in a reasonable worst-case scenario. This is simply `timeline_limit + len(required_state)`
   (ignoring `*` wildcards on state). Call it `M`.
 - If N > M then we would probably send more events if we did a delta than just telling the client
   everything from scratch, so issue a `SYNC` for this room.
 - If N < M then we don't have many events since the connection was last established, so just send
   the delta as an `UPDATE`.

This approach has numerous benefits:
 - In the common case when you scroll a room, you won't get any `SYNC`s for rooms that were invalidated
   because it's highly unlikely to receive 10+ events during the room scroll (assuming you scroll
   back up in reasonable time).
 - When you reconnect after sleeping your laptop overnight, most rooms will be `UPDATE`s, and busy

>    rooms like Matrix HQ will be `SYNC`ed from fresh rather than sending 100s of events.

Reviewer 9: Do I understand it right, that the client would get the complete state of the room again because auf `SYNC`? If yes: I don't see how that would consist to "bandwidth-efficient way". I have fully offline cabable clients in mind, which do not really use the windows API features, because they want to retrieve nearyl everything. The windows API may be good for a fast first login or always online clients, but for syncing just the difference to the previous sync it feels like a downgrade compared to sync v2. But maybe I just misunderstood, how sync v3 works :D

Author: It depends on the filters the client has applied. If the client is asking for all state events in the room (so for Matrix HQ this is something like 25k events) then the idea is that we will only start from fresh if you have missed >25k events in that room, else you will just get the delta.

   along with the stream position when that was sent. So it's basically a `map[string]int64`.

TODO: Name of opt-in flag? Configurable flags for 'I remember state' vs 'I remember timelines'? Do
we need to explicitly mark rooms as `delta: true` or `initial: true` or something to make it clear
whether this data should update/replace what the client knows?

### E2EE Handling

The server cannot calculate the `highlight_count` in E2EE rooms as it cannot read the message content.
This is a problem when clients want to sort by `highlight_count`. In comparison, the server can
calculate the name, `unread_count`, and work out the most recent timestamp when sorting by those
fields. What should the server do when the client wants to sort by `highlight_count` (which is pretty
typical!)? It can:
 - Assume `highlight_count == 1` whenever `unread_count > 0`. This ensures that E2EE rooms are always
   bumped above unreads in the list, but doesn't allow sorting within the list of highlighted rooms.
 - Assume `highlight_count == 0` always. This will always sort E2EE rooms below the highlight list,
   even if the E2EE room has a @mention.

>  - Sort E2EE rooms in their own dedicated list: `{"filters": { "is_encrypted": true }}`

Reviewer 6: I think it's the best idea. Considering the fact not every client supports E2EE anyway it may have other benefits.

Author: The API lets clients choose what they want to do here. For a lot of clients I suspect including them in the same list is probably okay, but certain clients may want that extra flexibility that a different list can support. Indeed, having the `is_encrypted` filter is particularly useful when I've been testing, as the test jig I wrote in the proxy didn't handle E2EE so it was nice to be able to filter them all out and thus save on bandwidth.

TODO: There is currently no lazy-loaded members support. Various suggestions like a special sentinel
value in required state e.g `["m.room.member", "LAZY"]` could be used. This may need to be specified
in this MSC or possibly an extension MSC (though it's unclear how it would fall under the filtering 
extension MSC format). For context, lazy-loaded members refers to the sending of `m.room.member`
events for the _senders_ of the events that are present in `timeline`. Current implementations use
a LRU cache to cut down on sending duplicate events. In addition, we may also want a sentinel value
to indicate "the current user" e.g to always pull out the current user's member event
`["m.room.member", "$current"]`. Are there any other interesting state filters?

### Sliding Window API

At a high level, the sliding window API provides a way to synchronise a subslice of a list in a
bandwidth efficient way. It does this by referring to "operations" which must be performed on the
stored client list, such as INSERT, DELETE and SYNC. Each operation has an index position OR a
range of index positions which tells the client where the operation should be performed. The

> possible operations are:

Reviewer 10: I wonder if it would make sense to use the JSON Patch RFC (https://www.rfc-editor.org/rfc/rfc6902) as base for the operations? Out of the box at least `INSERT` and `DELETE` should work, I guess.

Reviewer 10: As an real life example of someone using the RFC: FHIR is using it for their PATCH endpoint (https://www.hl7.org/fhir/http.html#patch).

Author: Relevant snippet:
```
[A.7](https://www.rfc-editor.org/rfc/rfc6902#appendix-A.7).  Moving an Array Element

   An example target JSON document:

   { "foo": [ "all", "grass", "cows", "eat" ] }

   A JSON Patch document:

   [
     { "op": "move", "from": "/foo/1", "path": "/foo/3" }
   ]

   The resulting JSON document:

   { "foo": [ "all", "cows", "eat", "grass" ] }
```

This naturally has more boilerplate, and it's unclear how it would handle sliding windows. This MSC uses absolute index positions when issuing move operations, which wouldn't work here as you would actually need an array of that size. You could add an `offset` integer to indicate where the array begins in the room list then use relative indexing, but now you need to maintain and think about 2 kinds of index: absolute and relative. It ends up feeling more complicated, and the value in RFC6902 is only there if there are a variety of good libraries in different languages so people don't need to implement this logic themselves.

        // events in these spaces will be returned. Unknown spaces or spaces the user is not joined to
        // will be ignored.
        "spaces": ["!foo:bar", "!bar:baz"],
        // Flag which only returns rooms which have an `m.room.encryption` state event. If unset,
        // both encrypted and unencrypted rooms are returned. If false, only unencrypted rooms
        // are returned. If true, only encrypted rooms are returned.
        "is_encrypted": true,
        // Flag which only returns rooms the user is currently invited to. If unset, both invited
        // and joined rooms are returned. If false, no invited rooms are returned. If true, only
        // invited rooms are returned.
        "is_invite": true,
        // If specified, only rooms where the `m.room.create` event has a `type` matching one
        // of the strings in this array will be returned. If this field is unset, all rooms are
        // returned regardless of type. This can be used to get the initial set of spaces for an account.
        // For rooms which do not have a room type, use 'null' to include them.

>         "room_types": [ "m.space", null ],

Reviewer 11: I was thinking about the `filters` again, in particular that someone might want to filter for certain subtypes (as [per spec MSC-3088](https://github.com/matrix-org/matrix-spec-proposals/pull/3088)) or other custom state, like having [`m.room.pinned_events`](https://spec.matrix.org/v1.4/client-server-api/#mroompinned_events) set, and notice this isn't possible at the moment.

We have support for some of these markers, like space, dm or encrypted, but not for others (even those that are in the spec, like pinned_events), which kinda feels odd. I am wondering if we can unify them to a general `has_state` and `not_has_state` with a similar matching systems as `required_states`. We could still keep the existing filters as we have them now, but allow for more generic (unknown to us and to the future of the spec) state-matching. What do you think?

Author: Arbitrary state filtering isn't great because it is hard to do this efficiently. Currently, the proxy pre-calculates values like DM-ness, encrypted-ness, spaces (child and parent) so it's very quick to then filter in-memory. Allowing arbitrary state filtering means we either need to do a DB hit (slow) or pre-cache all state in-memory (expensive and basically unscalable).

I can see the value in the flexibility, but not at the cost of performance imo which is the entire point of this MSC after all.

`/sync` v2's `since` query parameter. This is called `pos` and represents the position in the stream
the client is currently at. Unlike `/sync` v2, this token is ephemeral and can be invalidated at any
time. When a client first connects to the server, no `pos` is specified. Also unlike `/sync` v2, this
token cannot be used with other APIs such as `/messages` or `/keys/changes`. Note that the "connection"
formed to the server is _not_ a long-lived TCP connection, it is just an application-level concept
of a connection.

In simple servers, the `pos` may be an incrementing integer, but more complex servers may use vector
clocks or contain node identifying information in the token. Clients MUST treat `pos` as an opaque
value and not introspect it.

When a `pos` is invalidated and the client attempts to use the `pos`, the server MUST send back a
standard error response as a HTTP 400 containing:
```js
{
  "error": "Unknown position",
  "errcode": "M_UNKNOWN_POS"
}
```
This then allows the client to reset their connection and send an initial request (with all sticky
request parameters) without a `pos` value to restart the connection.

> The `timeout` query parameter exists for the same purposes of sync v2: to tell the server how many

Reviewer 2: we should specify a default value, no?

        }
    }
}
```
Extensions MUST have an `enabled` flag which defaults to `false`. If a client sends an unknown extension
name, the server MUST ignore it (or else backwards compatibility between servers is broken when a newer
client tries to communicate with an older server). Extension args may or may not be sticky, it
depends on the extension.

Extensions can leverage the data from the core API, notably which rooms are currently inside sliding
windows as well as which rooms are explicitly subscribed to.

In an effort to reduce the size of this proposal, extensions will be done in separate MSCs. There will
be extensions for:
 - To Device Messaging - [MSC3885](https://github.com/matrix-org/matrix-spec-proposals/pull/3885)
 - End-to-End Encryption - [MSC3884](https://github.com/matrix-org/matrix-spec-proposals/pull/3884)
 - Typing Notifications - [impl](https://github.com/matrix-org/sliding-sync/blob/main/sync3/extensions/typing.go)
 - Receipts - [impl](https://github.com/matrix-org/sliding-sync/blob/main/sync3/extensions/receipts.go)
 - Presence - `presence` in sync v2: [spec](https://spec.matrix.org/v1.1/client-server-api/#get_matrixclientv3sync)

>  - Account Data - `account_data` in sync v2:  [impl](https://github.com/matrix-org/sliding-sync/blob/main/sync3/extensions/account_data.go)

Reviewer 2: the current account data extension is distinctly problematic, in that without a `since` token, you're guaranteed to receive all your entire account data every time your session expires (i.e. on pretty much every app launch, after 30m of downtime). For me, that's 150KB of dense JSON that has to be calculated, transferred and decoded before the app can launch.  Of this, 22KB is push rules and 116KB is tracking which rooms are tagged as DMs.  Empirically this seems to delay launch very noticeably relative to before we had the account_data extension.

```

Reviewer 2: (delta tokens are the proposed solution to this)

_bi-directional_ exchange of JSON objects. This mechanism is ideally suited for WebSockets, but more
difficult to do for HTTP long-polling. This design was chosen in order to allow for a seamless transition
to a stream-orientated protocol like WebSockets in the future.

The existing `/sync` implementation in Matrix also creates a stream but it has limitations. It uses
a `since` token to tell clients where in the stream they are and to tell servers which messages the
client has received (in other words it serves as an ACK). Critically, the stream is _not stateful_. The
request must contain the entire set of input parameters, either via a filter ID or in-line filter. This
results in clients using the same set of input parameters most of the time. In order for sliding sync
to provide only the data needed to render the UI and nothing more, the set of input parameters needs to
be greatly expanded and they need to be dynamic: adding and removing parameters on-the-fly, without
additional round trips. In order to achieve this, Sliding Sync creates _stateful_ connections to the
server, so clients can simply send the deltas. This means clients and servers need to have a mechanism

> to agree on what that stored state is. _This introduces additional rules on client implementations._

Reviewer 7: Does this mean that sync requests actually are tracked by HTTP connection? And I can't have multiple sync streams without figuring out how to not make my library multiplex those streams onto the same connection? That doesn't sound that desirable. Ideally I would be able to handle presence and to_device messages at a different pace than room updates. The obvious implementation for that would be to open multiple streams. With the current sync that can't be done, since it deletes your to_device messages, even when you tried to filter them out (at least on some servers). However that is already solved for sliding sync.

So passing along some stream id, possibly a single character, would allow for multiple parallel streams at different frequencies as well as fast reconnect to old streams after a connection interruption (which can happen quite a lot on lossy networks or when switching networks without http/3). You kinda need a stream id, since you don't have a since token here to distinguish streams.

Reviewer 7: (For reference, presence accounts for about 80% of the CPU usage on Nheko, since most of the time it is presence events, that wake up the sync loop.)

Author: Currently, the proxy will track connections by the access token used in the request, forcing 1 connection per 1 access token.

This is absolutely something we're considering, because it also has a negative impact on Web, whereby if you open a SS-enabled client in two browser tabs they step on each other and expire each other's connnections. Any potential stream ID needs to be client-driven, as I don't want 5x `/sync` requests without a `?pos=` to cause 5 connections to be made, i.e the stream ID needs to be explicitly in the client request, not generated server-side.

In a WebSockets implementation, this is easy: the request parameters are sent initially when a connection
is established and then remain active for the lifetime of the connection. Any changes to these parameters
are reliably sent to the server in the order they were submitted. In order for clients to know when these
parameters have been applied, most application-level WebSocket protocols use a "message ID" chosen by the
client which is then echoed back in the ACK message. This is very similar to transaction IDs on the `/send`
endpoint in Matrix. 

However, this proposal does not use WebSockets; it uses HTTP long-polling. Like with `/sync`, this proposal
uses a token to allow servers to know which messages the client has received. Emulating WebSockets over HTTP
long-polling is difficult and has limitations. Servers cannot push new data to the client and must instead
wait for the client to make an HTTP request. In addition, individual HTTP requests can fail, resulting in
ordering problems which simply do not exist in a WebSockets implementation. This can lead to some
counter-intuitive responses from a Sliding Sync enabled server, unless certain rules are followed.

> **Long-polling Rule 1:** do not send multiple concurrent sliding sync requests to the server. If a request is lost in

Author: The server can't distinguish this failure mode from a genuine 2nd concurrent request to change request params. This rule is focusing on the case where the request may or may not make it to the server.

  A,1    | --------B,1--/     |        A        1
  A,1    | --------C,1------> |        A        1
         |                    |        A,C      2
  A,1    |         /--------- |        A,C      2
         |                    |        A,C      2
  A,1    | --------D,1------> | 1 != 2 -> missed a response
         |                    |        A,C,D    3
A,C,D    | <------C,D,3------ |        A,C,D    3
```
At this point, the client knows that `B` never made it to the server, because C was _sent after_ B, and
the server has ACKed C. If requests were sent in parallel (B and C at the same time), it would be impossible
for clients to know if B was still processing or if B had failed entirely.

**Long-polling Rule 2:** use transaction IDs if you need to know when a response has been applied. The above example used
A,B,C,D as transaction IDs, but in reality requests/responses are not always obviously tied together. For example,
requesting the first 10 rooms on a users account may return 0 results or 10: it's not possible to know ahead of time.
Clients need to know this information to know when to stop showing a spinner for example. For these reasons, clients 

> SHOULD send a transaction ID when they need to know when the response has been calculated.

Reviewer 7: Why can't a client just tell something has been responded to by looking if the response is to their particular request? What http library does not tie a response to a request handle?

Author: Because of rule #3 - the HTTP response may not be for the HTTP request you made (if, for example, the response is live data).

it will send the most recent unacknowledged response, in this case C, _even though_ the HTTP request was for D.

#### Message IDs for clients and servers

For the long-polling use case, this proposal includes an opaque token that is very similar to
`/sync` v2's `since` query parameter. This is called `pos` and represents the position in the stream
the client is currently at. Unlike `/sync` v2, this token is ephemeral and can be invalidated at any
time. When a client first connects to the server, no `pos` is specified. Also unlike `/sync` v2, this
token cannot be used with other APIs such as `/messages` or `/keys/changes`. Note that the "connection"
formed to the server is _not_ a long-lived TCP connection, it is just an application-level concept
of a connection.

In simple servers, the `pos` may be an incrementing integer, but more complex servers may use vector
clocks or contain node identifying information in the token. Clients MUST treat `pos` as an opaque
value and not introspect it.

When a `pos` is invalidated and the client attempts to use the `pos`, the server MUST send back a

> standard error response as a HTTP 400 containing:

Reviewer 7: What happens if a client tries to use an older `pos` token? Will the connection rewind to that state or will it return an error? Does this include when the client asks for the same `pos` twice, because it lost the response?

Author: It will return an error. It does not happen when the client asks for the same pos twice because it lost the response.

The proxy implements this by holding onto 2 position values:
 - the position just sent back to the client i.e in `{ pos: ... }` <- Unacknowledged latest position
 - the position last received in a client request i.e in `?pos=` <- Acknowledged position

```js
{
    "foo": {
        "bar": 3
    }
}
```
What is the value of `baz`? Both unset and `2` are valid answers. For this reason, `baz` MUST
be marked as sticky if the desired result is `2`, else it will be unset.

In order for servers and clients to agree on the set of sticky parameters, clients MUST send a transaction
ID with each change to their request parameters and servers MUST buffer responses. This transaction ID will
be echoed back to the client so it knows that those parameters have been applied. If the request parameters
have not been modified, then the `txn_id` does not need to be sent.

### Room List parameters

One or more room lists can be requested in sliding sync like so:
```js
{
  // A map of list key to list information. Max lists: 100.
  "lists": {
    // an arbitrary string which the client is using to refer to this list for this connection. Keep
    // this small as it needs to be sent a lot. Max length: 64 bytes.
    "client_chosen_key": {
      // Sliding window ranges, see the Sliding Window API for more information.
      // If this field is missing, no sliding window is used and all rooms are returned in this list.

>       "ranges": [ [0,99] ],

Reviewer 7: Is this sticky? Possibly exclusive with the `slow_get_all_rooms`, so that I only specify the range once, sync 3 times, then when I want all rooms, I clobber it with `slow_get_all_rooms`?

Author: It's not sticky in the MSC, but I'd need to check if the proxy treats this as sticky. It used to not, but this was before we swapped to lists-as-keys.

      // are changed. Elements in this array are ORd together to produce the final set of state events
      // to return. One unique exception is when you request all state events via ["*", "*"]. When used,
      // all state events are returned by default, and additional entries FILTER OUT the returned set
      // of state events. These additional entries cannot use '*' themselves.
      // For example, ["*", "*"], ["m.room.member", "@alice:example.com"] will _exclude_ every m.room.member
      // event _except_ for @alice:example.com, and include every other state event.
      // In addition, ["*", "*"], ["m.space.child", "*"] is an error, the m.space.child filter is not
      // required as it would have been returned anyway.
      "required_state": [
        // Request the join rules event. Note that the empty string is required here to match
        // the event's blank state_key.
        ["m.room.join_rules", ""],
        ["m.room.history_visibility", ""],
        // Request all `m.space.child` state events.
        // The * is a special sentinel value meaning 'all keys'.
        // Note that `*` is NOT a generic glob function. You cannot specify `foo*` to pull in keys

>         // like `food` and `foobar`. In this case, the * is treated as a literal *.

Author: I think there is value to state what `*` means in this scenario. The previous sentences merely state that:
 - it isn't a globbing function.
 - it won't pull in keys like...

Okay, so what _will_ it do? It's treated as a literal *.

        ["m.space.child", "*"],
        // Request only the m.room.member events required to render events in the timeline.
        // The "$LAZY" value is a special sentinel value meaning "lazy loading" and is only valid for
        // the "m.room.member" event type. For more information on the semantics, see "Lazy-Loading Room Members".
        ["m.room.member", "$LAZY"],
        // Request your own m.room.member event.
        // The "$ME" value is a special sentinel value meaning "my user id". It is valid for use on
        // any state event, but is typically most useful on the m.room.member event.

>         ["m.room.member", "$ME"],

Author: It's very convenient to have this in Element-Web, as I can define the lists as `const`s by using `$ME`, whereas without it I have to keep the variable in sync with whatever `MatrixClient` is hanging around (and remake it when it changes).

        // Request all state events.
        ["*", "*"]
      ],
      // Sticky. The maximum number of timeline events to return per response.

>       "timeline_limit": 10,

Author: The server should be able to not honour that i.e a newly joined room over federation isn't going to cause tons of backfilling because you want many timeline events. I suspect the semantics are going to be the same as sync v2 filters.

      // See the "Tombstones" section for more information.

>       "include_old_rooms": { //sticky

Author: Yes, it's a shallow replace. If this object is provided, it replaces the object that came before it.

The "ambiguous api" section describes the behaviour when this is shown, which states:

> What is the value of `baz`? Both unset and `2` are valid answers. For this reason, `baz` MUST
be marked as sticky if the desired result is `2`, else it will be unset.

"else it will be unset" i.e everything in this object will be unset if it isn't marked as sticky.

>       "include_old_rooms": { //sticky

Reviewer 12: Would it be worth eliminating the ambiguity here and stating the whole object is sticky, then?

        "timeline_limit": 1,
        "required_state": [ ["m.room.tombstone", ""] ]
      },

      // Sticky. Filters to apply to the list before sorting.
      "filters": {
        // All fields below are Sticky.
        // All fields are applied with AND operators, hence if is_dm:true and is_encrypted:true
        // then only Encrypted DM rooms will be returned. The absence of fields implies no filter
        // on that criteria: it does NOT imply 'false'.
        // These fields may be expanded through use of extensions.

        // Flag which only returns rooms present (or not) in the DM section of account data.
        // If unset, both DM rooms and non-DM rooms are returned. If false, only non-DM rooms
        // are returned. If true, only DM rooms are returned.
        "is_dm": true,

>         // A list of spaces which target rooms must be a part of, as m.space.child state events.

Reviewer 7: What about hidden space children? I.e. when a space admin only sets the m.space.parent event in a room, so that space members can only know about the room, when they are joined to the child room? Those should be listed in the /hierarchy call, so I am suprised, that those seem to be excluded here.

        // The server will inspect the m.space.child state events for the JOINED space room IDs given,
        // and filter the room list based on the INVITED/JOINED children room IDs.
        // If the child room has a m.room.tombstone event, then the search should recursively navigate
        // the room ID in that event to find the latest room and use that room ID instead of the initial
        // room ID in the m.space.child event.
        // If unset, all rooms are included. Servers MUST NOT navigate subspaces. It is up to the client to
        // give a complete list of spaces to navigate. Only rooms directly mentioned as m.space.child
        // events in these spaces will be returned. Unknown spaces or spaces the user is not joined to

>         // will be ignored.

Author: Yes and Yes.

Assuming the client is joined to spaces `["!a:example.org", "!b:example.org"]` then a room can pass the filter if the room belongs to ANY one of the spaces given.

        "spaces": ["!foo:bar", "!bar:baz"],
        // Flag which only returns rooms which have an `m.room.encryption` state event. If unset,
        // both encrypted and unencrypted rooms are returned. If false, only unencrypted rooms

>         // are returned. If true, only encrypted rooms are returned.

Author: Simply present, else we're baking in too much in case the event sprouts a new algorithm/structure.

        "is_encrypted": true,
        // Flag which only returns rooms the user is currently invited to. If unset, both invited
        // and joined rooms are returned. If false, no invited rooms are returned. If true, only
        // invited rooms are returned.

>         "is_invite": true,

Reviewer 7: Slightly related, do I get an update for a room I am kicked from or have left? As in, can I show the kick reason locally? Do I get a DELETE for that room (even if I have the list expanded to show all rooms) or how does that work? Does that also need an "is_leave" filter, which then is ORed together?

>         "is_invite": true,

Author: There is no knocking support yet, so we'd probably add `is_knock`.

Please make a separate comment re rooms being present after kick. The proxy will keep the room in the list _if the kick happens whilst connected_, but will be omitted on the next connection. It's not specced currently, and I don't know if that is desirable behaviour.

exists to allow the ability to search by room name which most clients support, and is crucial for
large accounts. The `room_types` filters exist primarily to include/exclude spaces. A previous version
of this MSC expressed multiple lists as an array and not an object with client-chosen keys. This was
changed because using arrays had a few undesirable consequences: you couldn't just edit list #3, you
had to add stub lists in index positions 0,1,2 first, and likewise the response demanded stub responses
which always included the `count` field to pad out earlier lists to get to the list index that was modified.
In addition, it was unclear how to delete a list. Also, some clients would race on startup to create lists,
which would result in different index positions being allocated, which made it hard for client code to
then refer deterministically to specific lists. The workaround basically assigned static names to each
list which then mapped to an index position. By using an object, these issues disappear._

The server will then return a `rooms` key which have the following fields:

```js
{
  "rooms": {
    // the room ID
    "!foo:bar": {

>       "name": "The calculated room name",

Author: AFAIK, yes.

We may add in the ability to specify a `locale` key (maybe in account data?) which would be read by the SS proxy which could convert `Alice and Bob` to `Alice et Bob`. Seems doable anyway.

>       "name": "The calculated room name",

Reviewer 8: I'm wary of regressing our ability to have native language names given we have many users in non-English speaking countries. Is this something we have buy in from the relevant teams (including customer facing teams)?

While we *could* fix this by introducing locale concept... that is quite a big change I don't foresee us doing any time soon (unless its done as part of SS work).

>       "name": "The calculated room name",

Reviewer 11: I agree that trying to keep i10n out of the protocol seems like the better approach. In the [rust SDK we made it more explicit](https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk/enum.DisplayName.html#variant.Calculated) whether the name was calculated or set. We could take a similar approach here by having a smarter type handing down the actual information for the calculation. Something like (don't take that format to literal, it's from top of my head):

```
"calculated_name": "Kate Jackson, Farrah Fawcett and Jaclyn Smith"
"name_source": { "format": "dm", "heroes_names": ["Kate Jackson", "Farrah Fawcett", "Jaclyn Smith" }
```

      // Flag which is set when this is the first time the server is sending this data on this connection.
      // Clients can use this flag to replace or update their local state. When there is an update, servers
      // MUST omit this flag entirely and NOT send "initial":false as this is wasteful on bandwidth. The

>       // absence of this flag means 'false'.

Author: @ara4n and myself both feel it's a huge mistake to allow servers to optionally include redundant fields. It ends up with brittle client implementations - see the many cases where Dendrite/Conduit omit fields that Synapse does not and it breaks one of Web/iOS/Android.

        {"sender":"@alice:example.com","type":"m.room.history_visibility", "state_key":"", "content":{"history_visibility":"joined"}},
        {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!foo:example.com", "content":{"via":["example.com"]}},
        {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!bar:example.com", "content":{"via":["example.com"]}},
        {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!baz:example.com", "content":{"via":["example.com"]}}
      ],
      // Last event is most recent. Max timeline_limit events.
      "timeline": [
        {"sender":"@alice:example.com","type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"invite"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"C"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"D"}},
      ],

>       "is_dm": true, // field is absent on non-DM rooms

Author: I don't agree with this re https://github.com/matrix-org/matrix-spec-proposals/pull/3575/files#r1120029360

>       "invite_state": [ { type: "m.room.member" } ], // stripped state events, same as rooms.invite.$room_id.invite_state in sync v2, absent on joined/left rooms

Reviewer 7: Maybe rename this to `stripped_state` and share it with kocked rooms?

      "prev_batch": "t111_222_333", // same as sync v2
      "limited": true,              // same as sync v2
      "joined_count": 41,           // same as sync v2 m.joined_member_count

>       "invited_count": 1,           // same as sync v2 m.invited_member_count

Reviewer 7: What about the knocked member count?

      "notification_count": 54,     // same as sync v2
      "highlight_count": 3,          // same as sync v2
      // The number of timeline events which have just occurred and are not historical.
      // The last N events are 'live' and should be treated as such.
      // This is mostly useful to determine whether a given @mention event should make a noise or not.
      // Clients cannot rely solely on the absence of 'initial: true' to determine live events because
      // if a room not in the sliding window bumps into the window because of an @mention it will have
      // 'initial: true' yet contain a single live event (with potentially other old events in the timeline)
      "num_live": 1

>     }

Reviewer 7: Attaching to a random position, since I don't know where to put it:

What I can't figure out at all is: How do I know what rooms I am a member of, what rooms I am invited to, what rooms I have knocked on and what rooms I have left? Is my own member event always guaranteed to be included the first time I get a room? If not, putting it in a "membership" key or so might be more convenient (current sync has it on the top level, so it isn't an issue).

Author: It depends on your filters. You're only going to be sent rooms you've joined, been invited to, or have knocked on (when it gets specced). If you specify no filter, then all of them will be returned in the list. If you are getting all of the rooms in the list and need to care about the membership, then ask for that state event `["m.room.member", "$ME"]`.

Reviewer 7: In the past the member event hasn't been resent based on state resets, so the only thing one could trust used to be what section the room comes down on /sync. I am assuming clients can rely on getting the right member event even after state resets then? As in a state reset causes the current state event to be pushed? That should probably be spelled out explicitly on the MSC then.

"@alice changed their name from Alice to Alice2". Event duplication may be
reduced using Event ID -> Event maps in the response should this be a concern. The benefit of
returning the current state is that servers can cache the latest state to return the response more
quickly. If, instead, servers returned the state at the start of a timeline block, servers are forced
to either rewind this state (as clients will need to do) or worse, do an expensive database access to
request the state before an event. As clients can be at different points in the stream for a given
room, this would force servers to cache every possible room state. It's not practical for servers to
cache every single possible earlier state for each room._

### Sliding Window API

At a high level, the sliding window API provides a way to synchronise a subslice of a list in a
bandwidth efficient way. It does this by referring to "operations" which must be performed on the
stored client list, such as INSERT, DELETE and SYNC. Each operation has an index position OR a
range of index positions which tells the client where the operation should be performed. The
possible operations are:

>  - `SYNC`: Sets a *range* of entries. Clients SHOULD discard what they previous knew about entries in

Reviewer 9: I want to understand, why we need such bloatet API. This makes the implementation more complicated. Why not just send a list of room ids, when the list changes? _At this point I would like to apologize if this discussion has already taken place and I have overlooked it._

Author: Sending list deltas is inherently complicated, which is why it's more complex than sync v2 which just sent everything. If you want a similar behaviour to that, just use `slow_get_all_rooms: true` on your list and then you won't see any ops at all.

You can't just send a list of room IDs because you need to communicate at which index they start at. If you sent the entire list and were in 10,000 rooms, that's 10,000 strings every time a room changed position, which wastes bandwidth.

   this range.
 - `INSERT`: Sets a *single* entry. If the position is not empty then clients MUST move entries

>    to the left or the right depending on where the closest empty space is.

Reviewer 13: - we just could forbid this? like "the server MUST ensure that this is unique."
- An alternative to keeping track of empty space would be to always have it in the end, that is moving to the front on DELETE and to the back on INSERT.

 - `DELETE`: Remove a *single* entry. Often comes before an `INSERT` to allow entries to move places.
 - `INVALIDATE`: Remove a *range* of entries. Clients MAY persist the invalidated range for offline
   support, but they should be treated as empty when additional operations which concern indexes in

>    the range arrive from the server.

Reviewer 13: > slightly confused about why we need both `DELETE` and `INVALIDATE`.

if i understand it correctly, DELETE(i) is not INVALIDATE(i,i), as `DELETE(i); INSERT(r,j)` with `window[i] =r`  does not need to send the information about room `r` in the `rooms` section of the response.

but is there any difference between invalidated entries and entries, that are deleted and not inserted again in the same response? is the client required to remember them?

>    the range arrive from the server.

Reviewer 11: I think the idea of invalidated is for performance reasons... assume the user scrolls is rather large list, then all these above can be kept around as invalidated in particular if you have an bouncer that only updates them once the scrolling is clear where to land. Rather than showing _no_ information, you show _stale_ information...

  // over all lists, including room subscriptions. This means if a room appears in 2 lists, only
  // 1 entry is present.
  "rooms": {
    "!foo:bar": {
      "name": "The calculated room name",
      // Additional response parameters omitted as they are
      // unrelated to the semantics of the sliding window.
      // See previous section on room list parameters.
    },
    // ... 9 more items
  },
}
```

_Rationale: Prior versions of this MSC more tightly coupled room data and list operations. This
became a problem if you did not want to use sliding windows because the room data will be contained
within list operations you don't care about. Now that this data is split out, it is easy for clients
to opt-out of sliding window semantics entirely (the `ops` key just disappears). Furthermore, the
`rooms` map was originally split out to be per-list / per-room-subscription but this could cause
needless duplication if a room appeared in >1 list. Each list can have different parameters associated
with them (e.g `required_state`, `timeline_limit`) but these can be aggregated / UNION'd easily._

> The possible `sort` operations are:

Reviewer 4: Bots and other roomlist-less clients probably won't care about the position of rooms, so to save bandwidth we could define a `none` sort option, which suppresses operations from being sent server->client. If the client needs to know which rooms to expect (ie: what `SYNC` would normally provide), it can call `/joined_rooms` or another similar API.

Or possibly we don't exclude `SYNC` from the sort, but do suppress everything else.

Reviewer 11: for bots and other roomlist-less clients, the `slow_get_all_rooms` was added with ignores any `sort` and `range`-parameters. Filter still apply though, so you can take the subset you care about (e.g. only DM's). Is there a use case that this doesn't cover yet?

Reviewer 4: I hadn't read that far in the MSC yet, and it didn't come up in discussion when this comment was written :p 

Will take a look at the other field.

   followed by all other rooms. See the "E2EE Handling" section for more information. Rooms are not sorted within each level:
   use an additional sort operation like `by_recency` to sort these groups. TODO: should we include
   unread indicator with this?
 - `by_name`: Sort by room name lexicographically. This requires servers to implement the
   [room name calculation algorithm](https://matrix.org/docs/spec/client_server/latest#calculating-the-display-name-for-a-room).
   The server MUST perform the following steps:
     * Calculate the room name from this user's perspective. This may vary depending on the user as
       DM rooms will have the room name set to the name of the _other user_. This is the value that
       will be returned in the `name` field for the room but is NOT the value that the server should
       perform sort operations on. See following steps.
     * Remove any of the following characters from the beginning/end of the calculated name: `#!()):_@`.
       This ensures things like canonical aliases display in roughly the right alphabetical locations
       rather than all together with all rooms that start with `#`.

>      * Lower-case the result by unicode. This ensures `Matrix` and `matrix` sort in the same locations.

Reviewer 14: I know you mention i18n below, but just to call out: "Lower-case the result by unicode" is wildly insufficient here. Different languages have completely different sorting preferences and it's basically only in English that sorting individual characters by their unicode codepoints will give anything approaching a sensible ordering.

Admittedly we're terrible at this today, so probably not a thing we need to fix in the first implementation.

     * Perform sort operations on this 'canonicalised' name. For clarity, the sort is descending so
       `A` comes before `B`.

Sorting algorithms MUST be stable and deterministic to avoid needless churn as otherwise identical
rooms keep swapping positions. This can easily be achieved by including a final tiebreak based on the
room ID (e.g lexicographical sort on the room ID) to guarantee stability and determinism. It is currently
_not possible_ to invert the sort order (ASC vs DESC). This may be added to this MSC if there is a
community need for it.

NOTE: It is known that by forcing servers to calculate the room name there can be problems concerning
multiple languages. "Alice and Bob" in English vs "Alice et Bob" in French for example, which may affect
sort ordering. This can be mitigated by adding a `lang` sticky request parameter to control how

> i18n and l10n are done.

Reviewer 14: to be explicit: this is proposed for a future MSC, rather than being part of this one?

  },
  "rooms": {
    "!encrypted:bar": {
      ...
    },
    // ... 1336 more items
    "!unencrypted:bar": {
      ...
    },
    // ... 19 more items
  },
}
```

### Room Subscription API

Sometimes clients know exactly which room they want to get information about e.g by following a
permalink or by refreshing a webapp currently viewing a specific room. The sliding window API alone
is insufficient for this use case because there's no way to say "please track this room explicitly".
The room subscription API serves as a way to provide this tracking. At a high level, the client
provides a map of room ID to room list parameters and the server then returns the response in the
same format as the sliding window API, just without the operations/indexes.

To track a room `!sub1:bar`, the client would send the following request:
```js
{
  "room_subscriptions": { // sticky
      "!sub1:bar": { // sticky
          "required_state": [ ["*","*"] ],
          "timeline_limit": 50,
          "include_old_rooms": { // See the "Tombstones" section for more information.
            "timeline_limit": 1,
            "required_state": [ ["m.room.tombstone", ""] ]
          },
      }
  }
}
```

> This would return the following response:

Reviewer 14: It slightly feels like it will be easy for client authors to leak subscriptions and basically forget that they've subscribed to rooms.

I also think that, in the case of a bug which causes clients and servers to get out of sync as to which rooms are subscribed, it's going to be quite tricky to track down what's going on.

To mitigate both issues: should we return a list of the subscribed rooms?

*alphabetical order* (`by_name`), which is what some other chat clients do for example. In this case,
you don't care about the event unless the event is a "highlightable" event (e.g direct @mention).
The notifications API exists to provide a mechanism for clients to display "unread messages"
indicators on the room list at positions not currently inside a sliding window.

TODO:
- Unsure how much data to expose (probably index position + notif/highlight counts?). If we do _counts_
then we are doomed to send a response to a client every time an event is sent in a noisy room, which
seems rather wasteful. Perhaps make it configurable? @timokoesters mentions having approx counts to
avoid the churn e.g only two digits of precision (21 -> 21, but 1234 -> 1200), this fits UIs very nicely.
- This has not been fully specced yet because in practice most clients sort by recency so it's not
  urgent to include this. For clients who sort by name though, this is a show stopper.

### Bandwidth optimisations for persistent clients

The Sliding Sync API assumes that room data is deleted on the client when:
 - the room falls out of the sliding window or;
 - a window gets invalidated or;

>  - the session expires and a new session is created.

Reviewer 7: So I am assuming a client should just ignore all SYNC, INVALIDATE and DELETE commands in respect to deleting the state here? Does that mean in case of a state reset, I can rely on the server sending me an "initial": true in the room object to tell me to reset my local state?

           |    Room1,timeline[$E,$F]         |
           |  req_state=[$F],delta_token=Y    |
```

#### Limitations

The delta token will not work under the following scenarios:
 - The timeline is filtered in some way. Currently Sliding Sync provides no filtering mechanism
   for timeline events but it will in the future. Any filters need to be the same between sessions
   for the bandwidth optimisations to work at all.
 - `m.room.member` events are excluded from these calculations. Delta tokens map to a lot of
   data server-side. In an effort to bound the growth of this data, `m.room.member` events MAY be
   sent redundantly even if the client has been sent it before. This also reduces the chances of
   missing an `m.room.member` event, which would risk causing E2EE key issues as the client would fail
   to encrypt for the target room member.
 - The server does not need to remember delta tokens and the associated data forever. The server
   can expire this data whenever they want, which will result in more redundant information being sent
   to the client and a new delta token being generated. This MSC recommends that servers keep delta

>    tokens valid for at least 7 days.

Reviewer 7: This means any persisting client needs to actively scan all timeline events for duplicates, since it can never know, if the server forgot that it already sent those?

Reviewer 7: Ah, I guess a client can tell, that is the case, because the room will have "initial": true?

will allow sorting based on ciphertext: this is an active area of research in the computer science
field.

### Extensions
We anticipate that as more features land in Matrix, different kinds of data will also want to be synced
to clients. Sync v2 did not have any first-class support to opt-in to new data. Sliding Sync does have
support for this via "extensions". Extensions also allow this proposal to be broken up into more
manageable sections. Extensions are requested by the client in a dedicated `extensions` block:
```js
{
    "extensions": {
        "name_of_extension": { // sticky
            "enabled": true, // sticky
            "extension_arg": "value",
            "extension_arg_2": true
        }
    }
}
```
Extensions MUST have an `enabled` flag which defaults to `false`. If a client sends an unknown extension
name, the server MUST ignore it (or else backwards compatibility between servers is broken when a newer
client tries to communicate with an older server). Extension args may or may not be sticky, it
depends on the extension.

Extensions can leverage the data from the core API, notably which rooms are currently inside sliding

> windows as well as which rooms are explicitly subscribed to.

Reviewer 7: Can they also add data to the core API in the **response** or will the response from the extension be in a separate field? I.e. when adding room account data, is that returned in the existing rooms or under something like `{"extensions": { "account_data": ... } }`

   For example, if the malicious user can engineer which hero's display names are used when calculating
   the room name (say by joining/leaving fake accounts) then it's possible for those names to advertise
   spam or spell out offensive words.

This API presents new ways for clients to request complex operations which runs the risk of denial
of service attacks:
 - Complex or pathological filter/sort options (especially via extensions) may degrade performance
   on the server and client. This may affect other users on the server.
 - Excessively long lists, list keys, ranges, etc. Some limits are specified in this MSC to mitigate against this.

# Unstable prefix

Whilst this in MSC review the HTTP path will be `/_matrix/client/unstable/org.matrix.msc3575/sync`
with the intention of this eventually becoming (confusingly) `/_matrix/client/v4/sync`. As this is a
brand new endpoint, no other keys or fields need prefixing.

Homeservers can advertise support for a sliding sync proxy by adding the following to their
`/.well-known/matrix/client` config:
```json
{
    "org.matrix.msc3575.proxy": {
        "url": "https://slidingsync.proxy.url.here"
    }
}
```

> This allows servers to declare an "official" trusted proxy, rather than using other URLs which may be

Reviewer 11: While implementing this for the [ansible docker builder](https://github.com/spantaleev/matrix-docker-ansible-deploy/pull/2515#discussion_r1119910354), the question came up whether this mean to define the entire URL, where to find he sliding-sync-endpoint or just a proxy that provides the `/_matrix/client/unstable/org.matrix.msc3575/sync` endpoint. Is this URL meant to include the path or not?

run by malicious actors who want to steal the access token for users.

# Dependencies

There are no MSCs required for the core functionality to be implemented. Servers and clients need to
be spaces-aware for spaces filters. Extension MSCs will depend on this MSC for their core functionality.

# Implementation state

> [Proxy server](https://github.com/matrix-org/sliding-sync) (v0.99.0):

Reviewer 15: Since the proxy seemingly will stay around for a while before we have this in synapse and others, we currently have a hard requirement for well-known which seems not optimal.

Element Android (see: https://github.com/vector-im/element-x-android/issues/137 ) therefor requires a well-known file now, which not everyone might be able to add in some edge cases. 

Maybe it would make sense as part of the unstable handling of this MSC to define a standard location for the sync proxy as long as it is required for sliding sync to work to be able to not require well-known for sliding sync usage.

For example, requiring `sliding.<server_name>` as the standard location of the proxy to check if no well-known was found and using the well-known mainly when the proxy is deployed elsewhere. This would allow to not hardcode the well-known requirement and make it easier to set up while the sync proxy is around.

Reviewer 16: I'd be concerned about requiring a standard subdomain - what if someone's infrastructure is already using that domain for something else?

I'd rather just ask clients to fall back to `<c-s url>/_matrix/client/unstable/org.matrix.msc3575/sync` if they can't find a valid `org.matrix.msc3575.proxy` field in a client .well-known file.

Reviewer 15: Ah, yes, that makes sense. :) I wasn't sure if the proxy works on a subpath or not since I only saw it used on a subdomain so far.

to rewind state to work out historical display names. Clients who show historical display names
already need to rewind state by inspecting the `prev_content` of an event to display text like
"@alice changed their name from Alice to Alice2". Event duplication may be
reduced using Event ID -> Event maps in the response should this be a concern. The benefit of
returning the current state is that servers can cache the latest state to return the response more
quickly. If, instead, servers returned the state at the start of a timeline block, servers are forced
to either rewind this state (as clients will need to do) or worse, do an expensive database access to
request the state before an event. As clients can be at different points in the stream for a given
room, this would force servers to cache every possible room state. It's not practical for servers to
cache every single possible earlier state for each room._

### Sliding Window API

At a high level, the sliding window API provides a way to synchronise a subslice of a list in a
bandwidth efficient way. It does this by referring to "operations" which must be performed on the

> stored client list, such as INSERT, DELETE and SYNC. Each operation has an index position OR a

Reviewer 17: Why are the ops all-uppercase? The only other thing that's all-uppercase in the protocol are error codes.

Reviewer 7: I at lest appreciate that they are upper case on the MSC, since that makes it much easier to search for them. The final values might want to be shortened to single characters though.

Reviewer 17: I'd much prefer lowercase but not single character, like basically everywhere else.

        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"C"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"D"}},
      ],
      "prev_batch": "t111_222_333",
      "joined_count": 4,
      "invited_count": 0,
      "notification_count": 54,
      "highlight_count": 3
    },
    // ... 99 more items
  },

  // Extensions API
  "extensions": {},

  // Bandwidth optimisations, see "Bandwidth optimisations for persistent clients"
  "delta_token": "server-generated-string" // the delta token to use
}
```
These fields and their interactions are explained in the next few sections. This forms the core of
the API. Additional data can be returned via ["extensions"](#extensions).

### Connections

At a high level, the syncing mechanism creates a "connection" to the server to allow the
_bi-directional_ exchange of JSON objects. This mechanism is ideally suited for WebSockets, but more
difficult to do for HTTP long-polling. This design was chosen in order to allow for a seamless transition

> to a stream-orientated protocol like WebSockets in the future.

Author: Yes.

results in clients using the same set of input parameters most of the time. In order for sliding sync
to provide only the data needed to render the UI and nothing more, the set of input parameters needs to
be greatly expanded and they need to be dynamic: adding and removing parameters on-the-fly, without
additional round trips. In order to achieve this, Sliding Sync creates _stateful_ connections to the
server, so clients can simply send the deltas. This means clients and servers need to have a mechanism
to agree on what that stored state is. _This introduces additional rules on client implementations._

In a WebSockets implementation, this is easy: the request parameters are sent initially when a connection
is established and then remain active for the lifetime of the connection. Any changes to these parameters
are reliably sent to the server in the order they were submitted. In order for clients to know when these
parameters have been applied, most application-level WebSocket protocols use a "message ID" chosen by the
client which is then echoed back in the ACK message. This is very similar to transaction IDs on the `/send`

> endpoint in Matrix. 

Author: We can probably have a grace period on it yeah, similar to connections today.

However, this proposal does not use WebSockets; it uses HTTP long-polling. Like with `/sync`, this proposal
uses a token to allow servers to know which messages the client has received. Emulating WebSockets over HTTP
long-polling is difficult and has limitations. Servers cannot push new data to the client and must instead
wait for the client to make an HTTP request. In addition, individual HTTP requests can fail, resulting in
ordering problems which simply do not exist in a WebSockets implementation. This can lead to some
counter-intuitive responses from a Sliding Sync enabled server, unless certain rules are followed.

> **Long-polling Rule 1:** do not send multiple concurrent sliding sync requests to the server. If a request is lost in

Reviewer 8: I don't quite follow how this works in practice? If sending a request times out, what does the client do? Considering the timed out request might still have reached the server

Reviewer 11: Times out client-side or server-side? Server-side it means the server responds with an kinda-empty response, so the client is meant to process it and go ahead with the next request. On client side, it works the same as any other network error: restart the connection - or do that after some back-off interval depending on what your client timeout handler strategy is.

Author: This doesn't involve creating new TCP connections. This is specifically around application level sliding sync requests. You don't need to send all the keys again for a simple timeout, only if the connection is explicitly expired with the error code mentioned in this MSC. In practice this means, when you send a request:
 - if it times out, retry.
 - if the request params change e.g the client scrolls a list, interrupt the request and send a new one. If your client cannot interrupt requests you can just send a new one, but then you _should_ ignore the response to the earlier request if you get one back.

The main thing you're trying to avoid here is if you send two distinct requests (i.e different req params) with the same `?pos=` then it can become impossible to know what the server-side state will be, and you risk getting out of sync.

Reviewer 8: > * if the request params change e.g the client scrolls a list, interrupt the request and send a new one. If your client cannot interrupt requests you can just send a new one, but then you _should_ ignore the response to the earlier request if you get one back.

Doesn't that contradict:

> The main thing you're trying to avoid here is if you send two distinct requests (i.e different req params) with the same `?pos=` then it can become impossible to know what the server-side state will be, and you risk getting out of sync.

In that, if the client "cancels" the outbound request the server may already have processed it and sent the response down. The client will then send out another request with different req params with the same `pos`, as it is ignoring the response from the cancelled request?

sent in the same response. In practice, the server does not combine multiple responses into a single response. Instead,
it will send the most recent unacknowledged response, in this case C, _even though_ the HTTP request was for D.

#### Message IDs for clients and servers

For the long-polling use case, this proposal includes an opaque token that is very similar to
`/sync` v2's `since` query parameter. This is called `pos` and represents the position in the stream
the client is currently at. Unlike `/sync` v2, this token is ephemeral and can be invalidated at any
time. When a client first connects to the server, no `pos` is specified. Also unlike `/sync` v2, this
token cannot be used with other APIs such as `/messages` or `/keys/changes`. Note that the "connection"
formed to the server is _not_ a long-lived TCP connection, it is just an application-level concept
of a connection.

In simple servers, the `pos` may be an incrementing integer, but more complex servers may use vector
clocks or contain node identifying information in the token. Clients MUST treat `pos` as an opaque

> value and not introspect it.

Author: Potentially, yes it does. I would rather not add an extra piece of data (session ID) to client book keeping, which implies it probably will have to live in `pos`. The proxy currently won't allow >1 session per access token.

will step on each other's toes. The act of hitting /sync without a ?pos results in the first tab being
torn down connection wise. We may want to mux in implicit session IDs into the pos?_

#### Sticky request parameters

Request parameters can be "sticky". This means that their value is remembered across multiple requests.
Clients cannot choose which parameters are sticky, the API defines which parameters are sticky.
The lifetime of sticky request parameters are tied to a sync connection. When the connection is lost,
the request parameters are lost with it. This feature exists to allow clients to configure the sync
stream in a bandwidth-efficient way. For example, if all keys were sticky:
```
Client                         Server
  | ------{ "foo": "bar" }------> |  {"foo":"bar"}
  | <-------HTTP 200 OK---------- |
  | ------{ "baz": "quuz" }-----> | {"foo":"bar","baz":"quuz"}
  | <-------HTTP 200 OK---------- |
```
For complex nested data, APIs which include sticky parameters MUST indicate every sticky field to
avoid ambiguity. For example, an ambiguous API may state the following:
```json5
{
    "foo": { // sticky
        "bar": 1,
        "baz": 2
    }
}
```

> When this object is combined with an the additional object:

Reviewer 12: ```suggestion
When this object is combined with the additional object:
```

```json5
{
    "foo": {
        "bar": 3
    }
}
```
What is the value of `baz`? Both unset and `2` are valid answers. For this reason, `baz` MUST
be marked as sticky if the desired result is `2`, else it will be unset.

In order for servers and clients to agree on the set of sticky parameters, clients MUST send a transaction
ID with each change to their request parameters and servers MUST buffer responses. This transaction ID will
be echoed back to the client so it knows that those parameters have been applied. If the request parameters
have not been modified, then the `txn_id` does not need to be sent.

### Room List parameters

One or more room lists can be requested in sliding sync like so:
```json5
{
  // A map of list key to list information. Max lists: 100.
  "lists": {
    // an arbitrary string which the client is using to refer to this list for this connection. Keep
    // this small as it needs to be sent a lot. Max length: 64 bytes.
    "client_chosen_key": {
      // Sliding window ranges, see the Sliding Window API for more information.
      // If this field is missing, no sliding window is used and all rooms are returned in this list.

>       "ranges": [ [0,99] ],

Author: Potentially, though we'd need to make sure clients see that the high index isn't the end of the room list (i.e it is `< count`) and so know to page in more data if they need it. I feel this will complicate client impls.

>       "ranges": [ [0,99] ],

Reviewer 8: This is a bit less of a concern if we allow clients to request all rooms, but its a potential nasty DoS vector server side.

      // Sticky. List sort order. See Sliding Window API for more information.
      // These fields may be expanded through use of extensions.
      "sort": [ "by_notification_level", "by_recency" ],
      // Sticky. Required state for each room returned. An array of event type and state key tuples.
      // Note that elements of this array are NOT sticky so they must be specified in full when they

>       // are changed. Elements in this array are ORd together to produce the final set of state events

Author: Yes, as explained in https://github.com/matrix-org/matrix-spec-proposals/blob/kegan/sync-v3/proposals/3575-sync.md#commonalities-between-the-room-subscription-api-and-sliding-window-api

        // The term 'like' is inspired by SQL 'LIKE', and the text here is similar to '%foo%'.
        "room_name_like": "foo",
        // Filter the room based on its room tags. If multiple tags are present, a room can have
        // any one of the listed tags (OR'd).
        "tags": ["m.favourite"],
        // Filter the room based on its room tags. Takes priority over `tags`. For example, a room
        // with tags A and B with filters tags:[A] not_tags:[B] would NOT be included because not_tags
        // takes priority over `tags`. This filter is useful if your Rooms list does NOT include the
        // list of favourite rooms again.
        "not_tags": ["m.lowpriority"]
      }
    }
  },
  // Sticky. Allowlist of event types which should be considered recent activity
  // when sorting `by_recency`. By omitting event types from this field, clients
  // can ensure that uninteresting events (e.g. a profile rename) do not cause a
  // room to jump to the top of its list(s). Empty or omitted `bump_event_types`
  // have no effect—all events in a room will be considered recent activity.
  // NB: this is currently per-connection, not per-list.

>   "bump_event_types": [ "m.room.message", "m.room.encrypted" ],

Reviewer 18: From talking to clients:

- it would be useful to have this field configurable per-list
- it would be useful to  bump_event_types set to an empty list. @kegsay described a use case: clients want a room list that never bumps rooms on activity _unless_ you were directly pinged---in which case, the most recent notification bubbled to the top of the list).

Author: Unfortunately the client UX is rather particular around this, and I do have to agree with the points they make. Here's some examples:
 - You have a muted, low priority room like Matrix HQ. You don't want it to bump to the top of your room list at all, unless you get directly @mentioned. 
 - You have a normal public room like #dendrite - you don't want membership changes or reactions to bump this room to the top of the list.
 - You have a DM - you _do_ want membership changes, reactions, etc to bump this room to the top of the list.

I share your concerns around efficiency however. Currently the proxy only really handles the `m.room.message` type correctly, because of precisely the problem you outline. This is further complicated by extensible events which could easily be multiple of these things (a message and a reaction for example). Perhaps we can restrict this to message events (which would correctly ignore member changes but would include reactions) but there are UX tradeoffs then.

Reviewer 8: Hmm, interesting. Do you know if these are hypothetical examples or something we're actually planning to implement? Some of those we just can't implement for E2E rooms.

How bad is it if the client gets told about Matrix HQ style stuff and then don't bump it up the room list? I guess it takes up a slot, though that is less bad in the persistent case?

had to add stub lists in index positions 0,1,2 first, and likewise the response demanded stub responses
which always included the `count` field to pad out earlier lists to get to the list index that was modified.
In addition, it was unclear how to delete a list. Also, some clients would race on startup to create lists,
which would result in different index positions being allocated, which made it hard for client code to
then refer deterministically to specific lists. The workaround basically assigned static names to each
list which then mapped to an index position. By using an object, these issues disappear._

The server will then return a `rooms` key which have the following fields:

```json5
{
  "rooms": {
    // the room ID
    "!foo:bar": {
      "name": "The calculated room name",
      // Flag which is set when this is the first time the server is sending this data on this connection.
      // Clients can use this flag to replace or update their local state. When there is an update, servers
      // MUST omit this flag entirely and NOT send "initial":false as this is wasteful on bandwidth. The
      // absence of this flag means 'false'.

>       "initial": true,

Author: Agreed, that sounds like a better name imo.

nonsensical and potentially expensive operations and B) not produce the best sort order for specific
use cases in Matrix such as alias handling. That being said, having some mechanism to support
additional sort operations is useful, see the extensions section for more information._

The complete API shape for each operation is shown below (note the key names vary on the operation):

```jsonl
{
  "op": "DELETE",
  "index": 8
}

{
  "op": "INSERT",
  "index": 99,
  "room_id": "!foo:bar"
}

{
  "op": "INVALIDATE",
  "range": [100,199]
}

{
  "range": [100,117],
  "op": "SYNC",
  "room_ids": [
    // ... 18 room IDs
  ]
}
```

Note that clients will NOT be notified of any events or activity in rooms not in the sliding window.
This can be a problem for some use cases:
 - Following a permalink to a random room which is not in the window should be possible.
 - Receiving a direct @mention in a room not in the window should notify the client.

For the first of these issues, the sliding sync API exposes a "room subscription" API. For the second issue,
the sliding sync API exposes a "notifications" API.

> #### Requesting all rooms

Author: This mode of operation is specifically targeting bot use cases, so it's funny you mention this. There are other use cases for this as well though, notably for clients to pre-cache some data about all rooms e.g the room name and avatar, for fast local room searches (which is what EX does).

I feel an outright ban is impractical here, however I do think we can probably add a clause here to say:

> Servers are allowed to return all rooms over a series of responses rather than as a single response.

This addresses the concern about making no forward progress, and bounds the response size, without complicating clients. The pagination is effectively then the `?pos=` value and can only go in one direction. It would probably look something like (for 500 rooms):

 - Response 1: `SYNC 0,99` -> 100 `rooms` in top level map
 -  Response 2: `SYNC 100,199` -> 100 `rooms` in top level map
 -  Response 3: `SYNC 200,299` -> 100 `rooms` in top level map
 -  Response 4: `SYNC 300,399` -> 100 `rooms` in top level map
 -  Response 5: `SYNC 400,499` -> 100 `rooms` in top level map

old and hence ineligible for being returned in a list.

#### Lazy-Loading Room Members

Room members in a room can be lazily-loaded by requesting the special value `$LAZY` as the state key
for the `m.room.member` event type in the `required_state` filter:
```json5
{
  "required_state": [
    ["m.room.member", "$LAZY"] // activate lazy loading
  ]
}
```
At a high level, this can be thought of as requesting the `m.room.member` events for a set of unknown
user IDs. Typically, when you view a room, you want to retrieve all state events _except_ for `m.room.member`
events which you want to lazily load. To get this behaviour, clients can send the following:
```json5
{
  "required_state": [
    ["m.room.member", "$LAZY"], // activate lazy loading
    ["*", "*"] // request all state events _except_ for m.room.member events which are lazily loaded
  ]
}
```
Check the description of `required_state` for more information on this behaviour, as it is not specific to lazy-loading.

The server processes `$LAZY` according to the following rules:
 - Calculate the timeline entries that will be returned in this room.

>  - For each timeline entry, ensure the `m.room.member` event for the `sender` of the timeline event is included exactly once per user ID.

Reviewer 18: As well as timeline entries, should this be triggered by typing notifications? (So that clients have a displayname to display and not just an MXID.) This seems to be what the test in https://github.com/matrix-org/sliding-sync/issues/55 is doing.

Probably best done with a sentence or two to say that
- extensions can also trigger a lazy loading member entry in `required_state`,
- extensions should not send down a duplicate event during live streaming (i.e. incremental syncs?)

and spell out explicitly in https://github.com/matrix-org/matrix-spec-proposals/pull/3961 that typing extensions should support lazy loading members in particular.

event which moves the room into the top 5. In most "recent" sort orders a new event *will result* in
the 6th room bumping to the top of the list. A notable exception is when the rooms are sorted in
*alphabetical order* (`by_name`), which is what some other chat clients do for example. In this case,
you don't care about the event unless the event is a "highlightable" event (e.g direct @mention).
The notifications API exists to provide a mechanism for clients to display "unread messages"
indicators on the room list at positions not currently inside a sliding window.

TODO:
- Unsure how much data to expose (probably index position + notif/highlight counts?). If we do _counts_
then we are doomed to send a response to a client every time an event is sent in a noisy room, which
seems rather wasteful. Perhaps make it configurable? @timokoesters mentions having approx counts to
avoid the churn e.g only two digits of precision (21 -> 21, but 1234 -> 1200), this fits UIs very nicely.
- This has not been fully specced yet because in practice most clients sort by recency so it's not
  urgent to include this. For clients who sort by name though, this is a show stopper.

> ### Bandwidth optimisations for persistent clients

Author: > without a delta token we assume that when a room falls of the end of the list the client will delete the data, but if there is a delta token then clients will keep the copy of all rooms?

Your understanding is correct.

My concern with allowing dangling rooms (not in the subscribed window, but still remembered position wise) is that it can quickly get out of sync. If you get a new event in a room, in practice it doesn't always bump it to the top of the list. This means you could be getting events for rooms off the end of the list which you aren't aware of. The server then has to either:
 - resend all the data for that room when it reappears in your list
 - remember where it got up to and then send the deltas to you when it reappears in the list. <- this could be a huge delta.

> ### Bandwidth optimisations for persistent clients

Reviewer 8: I was more suggesting that we *stay* subscribed to the room rather than leaving it dangling, e.g. instead of `[A,B,C,D] -> [E,A,B,C]` we'd do `[A,B,C,D] -> [E,A,B,C,D]` until either the client or server specifically reduced the range again. The client would do it whenever it decided it wanted to reclaim some space, and the server could do it... whenever it wanted.

explicit room subscription. In this case, the extension should process that room if the extension
is configured to process any of the windows/subscriptions that contains the room. (The logic is a union
of conditions, not an intersection.)

Extensions SHOULD NOT attach their own semantics to the `lists` and `rooms` fields.
Extensions are otherwise free to define and process their own config fields, which may be sticky.
Such fields are ignored by the Core of sliding sync and transparently forwarded to extensions.

In an effort to reduce the size of this proposal, extensions will be done in separate MSCs. There will
be extensions for:
 - To Device Messaging - [MSC3885](https://github.com/matrix-org/matrix-spec-proposals/pull/3885)
 - End-to-End Encryption - [MSC3884](https://github.com/matrix-org/matrix-spec-proposals/pull/3884)
 - Typing Notifications - [MSC3961](https://github.com/matrix-org/matrix-spec-proposals/pull/3961)
 - Receipts - [MSC3960](https://github.com/matrix-org/matrix-spec-proposals/pull/3960)
 - Presence - `presence` in sync v2: [spec](https://spec.matrix.org/v1.1/client-server-api/#get_matrixclientv3sync)

>  - Account Data - `account_data` in sync v2:  [MSC3959](https://github.com/matrix-org/matrix-spec-proposals/pull/3950)

Reviewer 19: Wrong PR number in the URL :)
```suggestion
 - Account Data - `account_data` in sync v2:  [MSC3959](https://github.com/matrix-org/matrix-spec-proposals/pull/3959)
```

Reviewer 18: Thanks, I have sausage fingers.

   the real `origin_server_ts` value in the event.
 - Subtly adjusting the events in the room could adjust the calculated room name to be inappropriate.
   For example, if the malicious user can engineer which hero's display names are used when calculating
   the room name (say by joining/leaving fake accounts) then it's possible for those names to advertise
   spam or spell out offensive words.

This API presents new ways for clients to request complex operations which runs the risk of denial
of service attacks:
 - Complex or pathological filter/sort options (especially via extensions) may degrade performance
   on the server and client. This may affect other users on the server.
 - Excessively long lists, list keys, ranges, etc. Some limits are specified in this MSC to mitigate against this.

# Unstable prefix

Whilst this in MSC review the HTTP path will be `/_matrix/client/unstable/org.matrix.msc3575/sync`
with the intention of this eventually becoming (confusingly) `/_matrix/client/v4/sync`. As this is a
brand new endpoint, no other keys or fields need prefixing.

Homeservers can advertise support for a sliding sync proxy by adding the following to their

> `/.well-known/matrix/client` config:

Reviewer 8: This probably should have been done as part of `/_matrix/client/versions` or `/capabilities` maybe?

          {"sender":"@alice:example.com","type":"m.room.member", "state_key":"@bob:example.com", "content":{"membership":"join","displayname":"Bob"}},
          {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
          {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
        ],
        "limited": true,
        // ...
    }
  }
}
```

Any updates in this room would be returned in the same section of the sync response:

```json5
{
  "rooms": {
    "!sub1:bar": {
        "timeline": [
          {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"C"}},
        ]
    }
  }
}
```

Multiple rooms can be subscribed to by specifying additional keys in the room subscription map. If
a room is subscribed to multiple times, the _most recent_ subscription takes effect for the purposes
of `required_state` and `timeline_limit` filtering.

To unsubscribe from a room, the client needs to send a request with the room ID to unsubscribe from in
the `unsubscribe_rooms` array:
```json5
{
  "unsubscribe_rooms": [ "!sub1:bar" ]

> }

Reviewer 8: I'm a bit concerned that this API will make it easy for clients to "forget" to unsubscribe from rooms, and you'd end up with clients slowly racking up lots of room subscriptions. Personally, I'd be tempted to change this such that every request you need to specify which rooms the client wants to remain subscribed to :shrug:

 - More complex display operations such as showing summed total notification counts in spaces.
 - More complex space operations such as handling orphaned rooms and traversal of subspaces.
 - More complex bot requirements like knowing all rooms which have a certain custom state event in it,
   such that the presence of a state event becomes a filter.

It is expected that some of these use cases **will be supported as this MSC is iterated upon**. However,
it is likely that some of these use cases will not be supported in this MSC, but may be supported via
use of an extension MSC where applicable. Unfortunately, there may be some data flows which are genuinely
impossible to perform due to limitations of server-side operations (e.g if the data is encrypted). In
this case, clients will be forced to pull in all E2EE rooms to perform their data flows, which, whilst
slow, should still perform better than sync v2.

This MSC alone won't meet the needs of the entire ecosystem in terms of sorting/filtering/data returned
to the client. Extensions are a crucial part of this MSC to clearly define how the sync API can expand
with changing requirements.

# Alternatives

> There are two main alternatives to this proposal:

Reviewer 9: Wouldn't be one alternative to use a more generic and common protocol like GraphQL? This would allow to subscribe to different room lists or timelines, to filter data and more, too. There are also many libraries implementing GraphQL, so there is no need to force all SDKs, clients and servers to implement a completely new and very complex sync protocol.

I think it should at least be discussed instead of reinventing the wheel.

Author: Added https://github.com/matrix-org/matrix-spec-proposals/blob/kegan/sync-v3/proposals/3575-sync.md#graphql

 - The server has reached a memory limit for your connection and has expired it to reclaim memory.
 - The server which handled your last request is no longer running (e.g it was restarted) and it cannot calculate a response.

To handle expired connections, clients should send an initial request (with all sticky request parameters)
without a `pos` value to restart the connection.

#### Concurrent connections

There are three main reasons why a client may want to have >1 connection to the server open concurrently:
 - The client is a **browser**, and it should be possible to open the same client in multiple tabs without causing problems. Without concurrent connections, each tab would reset the other tabs connection due to different `?pos=` values being sent. The number of concurrent connections is technically unbounded.
 - The client is a **mobile application**, and it should be possible to have a "push process" connection in addition to the "app connection". Without concurrent connections, it isn't possible to obtain to-device messages in the push process, whilst also obtaining them in the main app. The number of concurrent connections is fixed e.g 2.

>  - The client wants to do a **one-shot request** for some data, without incurring latency/bandwidth penalties with all the activity on the user's account. Without concurrent connections, it isn't possible to get the response without also potentially getting large amounts of extraneous data. The number of concurrent connections is N+1, where N is the number of active concurrent connections.

Reviewer 20: > The number of concurrent connections is N+1, where N is the number of active concurrent connections.

What is this trying to convey (can someone re-word)? I'm reading this as "a dog is a dog"

        ["m.space.child", "*"],
        // Request only the m.room.member events required to render events in the timeline.
        // The "$LAZY" value is a special sentinel value meaning "lazy loading" and is only valid for
        // the "m.room.member" event type. For more information on the semantics, see "Lazy-Loading Room Members".
        ["m.room.member", "$LAZY"],
        // Request your own m.room.member event.
        // The "$ME" value is a special sentinel value meaning "my user id". It is valid for use on
        // any state event, but is typically most useful on the m.room.member event.
        ["m.room.member", "$ME"],
        // Request all state events.
        ["*", "*"]
      ],
      // Sticky. The maximum number of timeline events to return per response.
      "timeline_limit": 10,

      // See the "Tombstones" section for more information.
      "include_old_rooms": { //sticky
        "timeline_limit": 1,
        "required_state": [ ["m.room.tombstone", ""] ]
      },

      // Sticky. Return a stripped variant of membership events (containing `user_id` and optionally `avatar_url` and `displayname`)
      // for the users used to calculate the room name.

>       "include_heroes": true,

Reviewer 20: What use case is `include_heroes` useful for?

`heroes` will already be populated when the room doesn't have a `m.room.name` set. And you can specify `$LAZY`-loading room members to get relevant members in the room.

Reviewer 20: `heroes` are used to [calculate the room name](https://spec.matrix.org/v1.10/client-server-api/#calculating-the-display-name-for-a-room) so being deterministic is useful so the room name doesn't change that much. `heroes` are already automatically returned when the room doesn't have `m.room.name` set. Since this behavior is already built-in and automatic, this `include_heroes` option isn't necessary for room name calculation.

Are `heroes` used for anything else? If not, then this option isn't necessary

---

Displaying a few avatar bubbles to preview some of the room members could be useful:

![](https://github.com/matrix-org/matrix-spec-proposals/assets/558581/e1971e6a-cd3a-4192-b909-3ccf4112f248)

A similar effect could be achieved by `$LAZY`-loading room members and fetching enough timeline events. `heroes` could also work but it will always only be up-to 5 members and won't reflect who's currently chatting if it's a big room. You could also fetch all members in the room by using `required_state: [["m.room.member", "*"]]` but can be a bit burdensome to parse in very large rooms. This use case might be better served by a specific extension to be able to specify the number of members you want to fetch and how you want them sorted.

Reviewer 12: heroes are already automatically returned when the room doesn't have m.room.name set.

> Are heroes used for anything else? If not, then this option isn't necessary

They're exposed to the EX apps at least, which use those to compute a hash that serves to determine a unique color for a room's visual entry in the room list. I think that color will also match what EW displays, since it's based on the same algorithm. That computation (be it done by the SDK or the EX apps themselves) requires the heroes to *always* be present.

Reviewer 20: Thanks for sharing another reference @bnjbvr :slightly_smiling_face:  

> use those to compute a hash that serves to determine a unique color for a room's visual entry in the room list

Why not just hash against the room ID?

Feels like the room color would change over time as the `heroes` will shift around until 5 people are in the room and will shift as people leave the room.

As far as I can tell, Element Web just bases it off the Matrix ID (for users) and room ID (for rooms):

 - [`useIdColorHash(...)`](https://github.com/element-hq/compound-web/blob/main/src/components/Avatar/useIdColorHash.ts#L17-L31)
 - [`compound-web` -> `src/components/Avatar/Avatar.tsx`](https://github.com/element-hq/compound-web/blob/15eefea5f821d546e6ea687c6da4b10af53b6f81/src/components/Avatar/Avatar.tsx#L107)
 - [`matrix-react-sdk` -> `src/Avatar.ts`](https://github.com/matrix-org/matrix-react-sdk/blob/19f9f9856451a8e4cce6d313d19ca8aed4b5d6b4/src/Avatar.ts#L107C17-L127)

> That computation (be it done by the SDK or the EX apps themselves) requires the heroes to always be present.

For Sync v2, `heroes` is only present for [`join` rooms](https://spec.matrix.org/v1.10/client-server-api/#_matrixclientv3sync_joined-room). And [in Synapse](https://github.com/element-hq/synapse/blob/3232bc29826d030e62c034ec582c73532bc7e465/synapse/handlers/sync.py#L2920-L2957), they are only returned when lazy-loading and (there are membership changes or initial sync).

So the color will change yet again based on membership if `heroes` was taken into account.

Can you share a link to the code in ElementX for this?

Reviewer 12: cc @jmartinesp @pixlwave ^

Reviewer 21: > Why not just hash against the room ID?

We do this most of the time, but when the room is a 1:1 DM then we use the other member's user ID so that the room's placeholder avatar matches the user's placeholder avatar. Web does this too I believe.

> Displaying a few avatar bubbles to preview some of the room members could be useful:

This is the future plan, there designs for avatar clusters of max 4 people so the hero limit of 5 isn't an issue for us.

<img width="262" alt="Screenshot 2024-07-10 at 6 28 08 PM" src="https://github.com/matrix-org/matrix-spec-proposals/assets/6060466/b7a2911f-121b-4bd4-8f31-4a8573ce873f">

Reviewer 21: Not sure if you need code too, but the check is [here](https://github.com/element-hq/element-x-ios/blob/84e785367b942b601016ee539b14ce84d63d9fd3/ElementX/Sources/Services/Room/RoomSummary/RoomSummary.swift#L87-L93) and rendering [here](https://github.com/element-hq/element-x-ios/blob/develop/ElementX/Sources/Other/SwiftUI/Views/RoomAvatarImage.swift)

Reviewer 20: Thanks for the extra clarification!

It appears that `heroes` are only necessary for DMs (or group chats), aligning perfectly with the server's automatic behavior of providing `heroes` when a room name is not specified.

This further supports the argument that a separate `include_heroes` option is unnecessary.

Reviewer 21: Does that mean we wouldn't get the heroes for the "Soccer Club" example in the screenshot above?

Reviewer 12: (A small note: there are other clients than Element in the world, and we should also keep it flexible for these.)

Reviewer 21: Let me see if I can find out actual product requirements for this on our side.

      // Sticky. Filters to apply to the list before sorting.
      "filters": {
        // All fields below are Sticky.
        // All fields are applied with AND operators, hence if is_dm:true and is_encrypted:true
        // then only Encrypted DM rooms will be returned. The absence of fields implies no filter
        // on that criteria: it does NOT imply 'false'.
        // These fields may be expanded through use of extensions.

        // Flag which only returns rooms present (or not) in the DM section of account data.
        // If unset, both DM rooms and non-DM rooms are returned. If false, only non-DM rooms
        // are returned. If true, only DM rooms are returned.
        "is_dm": true,

>         // A list of spaces which target rooms must be a part of, as m.space.child state events.

Reviewer 20: Does this act like the `tags` filter below where a room can belong to any one of the spaces listed (OR'd)?

Reviewer 20: Answered at https://github.com/matrix-org/matrix-spec-proposals/pull/3575#discussion_r1113928699,

> Assuming the client is joined to spaces `["!a:example.org", "!b:example.org"]` then a room can pass the filter if the room belongs to ANY one of the spaces given.

>         // The server will inspect the m.space.child state events for the JOINED space room IDs given,

Reviewer 20: Should this also include public spaces that you haven't joined?

>         // The server will inspect the m.space.child state events for the JOINED space room IDs given,
>         // and filter the room list based on the INVITED/JOINED children room IDs.

Reviewer 20: The phrasing in this comment is a bit weird. This is in the context of a filter for the rooms in the sync response we already have, whereas it sounds like we're trying to discover rooms in the space.

Maybe this sounds more clear:

```suggestion
        // The server will inspect the `m.space.child` state events for the JOINED space room IDs given,
        // and filter the sync room list based on the space children room IDs listed.
```

        // If the child room has a m.room.tombstone event, then the search should recursively navigate
        // the room ID in that event to find the latest room and use that room ID instead of the initial

>         // room ID in the m.space.child event.

Reviewer 20: Why do we want to recursively find the latest room? What benefit for this extra work?

It seems like the new room would also be part of the space so it will be included anyway.

And tombstoned rooms would be filtered out depending on `include_old_rooms` (see *Tombstones* section as well).

Reviewer 22: > It seems like the new room would also be part of the space.

Only if the user upgrading had permissions to manage the space (and was on a client e.g. EW/ED which had this logic) - quite a lot of spaces are soft-broken where they only know about old versions of rooms.

Reviewer 20: Seems like if the spaces still know about the old rooms, people will visit the old room, be pointed to the new one and can ask the admin to fix space/room relationship.

Reviewer 22: @MadLittleMods sure, except the space admin might not even be in the room for them to as ask. So they'll likely end up asking the wrong admin.

Reviewer 20: Kinda feels like the same situation as a song playlist on Spotify/YouTube, etc. I'm not blaming the new song for not being in the playlist yet, I'm going to the playlist author to fix it.

Do we try to resolve room upgrades anywhere else?

>         // If the child room has a m.room.tombstone event, then the search should recursively navigate
>         // the room ID in that event to find the latest room and use that room ID instead of the initial
>         // room ID in the m.space.child event.

Reviewer 20: Same reasoning as the [other comment](https://github.com/matrix-org/matrix-spec-proposals/pull/3575#discussion_r1621361512).

> The phrasing in this comment is a bit weird. This is in the context of a filter for the rooms in the sync response we already have, whereas it sounds like we're trying to discover rooms in the space.
> 
> Maybe this sounds more clear:

```suggestion
        // If the child room has a `m.room.tombstone` event, then the search should recursively navigate
        // until we find the the latest room and consider those room IDs as part of the space.
```

exists to allow the ability to search by room name which most clients support, and is crucial for
large accounts. The `room_types` filters exist primarily to include/exclude spaces. A previous version
of this MSC expressed multiple lists as an array and not an object with client-chosen keys. This was
changed because using arrays had a few undesirable consequences: you couldn't just edit list #3, you
had to add stub lists in index positions 0,1,2 first, and likewise the response demanded stub responses
which always included the `count` field to pad out earlier lists to get to the list index that was modified.
In addition, it was unclear how to delete a list. Also, some clients would race on startup to create lists,
which would result in different index positions being allocated, which made it hard for client code to
then refer deterministically to specific lists. The workaround basically assigned static names to each
list which then mapped to an index position. By using an object, these issues disappear._

The server will then return a `rooms` key which have the following fields:

```json5
{
  "rooms": {
    // the room ID
    "!foo:bar": {

>       "name": "The calculated room name",

Reviewer 20: To preface this: I'm on the side of not trying to calculate room names. Just leave it to clients to use the explicitly set room name or calculate it themselves via the `heroes`.

One thing, I'm not sure about is if the `name` field should strictly be the value from the `m.room.name` state event or whether we should also fallback to the `m.room.canonical_alias`. The existing spec currently says that `heroes` should be returned ["if the room’s `m.room.name` or `m.room.canonical_alias` state events are unset or empty."](https://spec.matrix.org/v1.10/client-server-api/#_matrixclientv3sync_roomsummary) so it might make sense to fallback.

Reviewer 12: Agreed; calculing room names server-side is doomed to cause incorrect i10n (as a room name computed from heroes would look, for all languages, like "person A **and** person B"), and closes the door to customization options (even though that argument is feeble; the user value in being able to choose how many heroes are used for a fallback calculated room name would likely be quite low).

      // Optional, nullable string: the MXC URL of the room's avatar. If omitted,
      // there is no change to the avatar. If null, the room now has no avatar.
      "avatar": "mxc://...",
      // Optional. If omitted there is no change to the heroes or the `name` was not
      // calculated using room heroes. `avatar_url` and `displayname` are optional.
      "heroes": [

>         {"user_id":"@alice:example.com","displayname":"Alice","avatar_url":"mxc://..."},

Reviewer 20: Should we just use the existing definition of [stripped state](https://spec.matrix.org/v1.10/client-server-api/#stripped-state)?

> Stripped state events can only have the `sender`, `type`, `state_key` and `content` properties present.

>       ],

Reviewer 20: Instead of returning stripped state for `heroes`, we could simply return the User ID's of the heroes like Sync v2 and return the full membership event in `required_state` like we're already doing for `$LAZY`-loading room memberships.

Reviewer 20: Conversation continued in https://github.com/matrix-org/matrix-spec-proposals/pull/4186#discussion_r1745962353

        {"sender":"@alice:example.com","type":"m.room.history_visibility", "state_key":"", "content":{"history_visibility":"joined"}},
        {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!foo:example.com", "content":{"via":["example.com"]}},
        {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!bar:example.com", "content":{"via":["example.com"]}},
        {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!baz:example.com", "content":{"via":["example.com"]}}
      ],
      // Last event is most recent. Max timeline_limit events.
      "timeline": [
        {"sender":"@alice:example.com","type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"invite"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"C"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"D"}},

>       ],

Reviewer 20: What should happen for `invite` rooms?

If we copy Sync v2, then the invite room will only have `invite_state` (no `timeline` events, `limited`, `prev_batch` token, or `required_state`).

Perhaps we could make the `rooms` more uniform regardless of membership though. The hard case is how these fields should be set in cases where the user's room membership is `invite` with `shared`/`joined` history visibility where they can't see any timeline events until they join.

**Where should the `prev_batch` token be pointing to?** If we do it just like other rooms, we paginate up to the `timeline_limit`, then all of those events are filtered out (because we're not joined yet) and the `prev_batch` token points to a place before those events that we never saw. And then potentially when we `join`, we will miss those events depending on what the client does.

If we change the scenario to a room with `world_readable` or `invited` history visibility where timeline events would be returned when you are invited; The answer is timeline events are returned, `limited` is set respectively, and `prev_batch` points to the first event to continue paginating (no problems there).

For `ban`/`leave`, we can just query for timeline events up to the point where they were banned/left (no problem here either).

---

*First discussed in https://github.com/element-hq/synapse/pull/17320#discussion_r1646581077*

Reviewer 20: I think I found the distinction for why we don't do this: local vs remote room

For a room where another local user is in the room, this kind of thing makes sense to me. Being `invited` to a room probably means you should have access to the room (access to the full state of the room and events based on history visibility) (no security concern). Stripped state makes more sense for `knock` rooms where you don't have access yet.

But the problem comes when you're the first one on your server to be invited to the room over federation to a remote room. When you [receive a federated invite](https://spec.matrix.org/v1.10/server-server-api/#inviting-to-a-room), you only have access to `event.unsigned.invite_room_state` so that's all that you can return to the client. You would have to go through the [`make_join`/`send_join` process](https://spec.matrix.org/v1.10/server-server-api/#joining-rooms) to start participating in the room and query more info.

So it makes sense just to do the same thing for all `invite` rooms, just return stripped state.

Reviewer 20: Related to https://github.com/matrix-org/matrix-spec/issues/380. As an example, when someone is invited to a room, we don't even have the `heroes` necessary to calculate the room name if the room doesn't have `m.room.name` set.

          "op": "SYNC",
          // Which index positions are affected by this operation. These are both inclusive.
          "range": [0, 9],
          // Which room IDs are affected by this operation. These IDs match up to the positions
          // in the `range`, so the last room ID in this list matches the 9th index. The room data
          // is held in a separate object.
          "room_ids": [
            "!foo:bar", // ... 9 more room IDs
          ]
        }
      ]
    }
  },
  // The room data to use for each room ID. This data represents the point in time AFTER all
  // ops have been applied. For example, if a room had 2 new events which changed its list position
  // then you could see `ops` with DELETE[4,!foo:bar], INSERT[0,!foo:bar], DELETE[0,!foo:bar], INSERT[1,!foo:bar]
  // then the room !foo:bar in this map MUST contain both events.
  //
  // This map will only contain rooms which are present in the list `ops` above. If there are no
  // `ops` (because there are no `ranges`) then all rooms which match the list filters will be
  // present in this list, unordered. This functionality is useful for clients which do not want

>   // to use sliding list semantics. This map is an aggregation of all rooms which can be returned

Reviewer 20: > If there are no`ops` (because there are no `ranges`) then all rooms which match the list filters will be present in this list, unordered. This functionality is useful for clients which do not want to use sliding list semantics.

This seems leftover before `slow_get_all_rooms` was added

          ]
        }
      ]
    },
    "list_unencrypted": {
      "count": 420,
      "ops": [
        {
          "op": "SYNC",
          "range": [0, 19],
          "room_ids": [
            "!unencrypted:bar", // ... 19 more room IDs
          ]
        }
      ]
    }
  },
  "rooms": {
    "!encrypted:bar": {
      // ...
    },
    // ... 1336 more items
    "!unencrypted:bar": {
      // ...
    },
    // ... 19 more items
  },
}
```

### Room Subscription API

Sometimes clients know exactly which room they want to get information about e.g by following a
permalink or by refreshing a webapp currently viewing a specific room. The sliding window API alone
is insufficient for this use case because there's no way to say "please track this room explicitly".
The room subscription API serves as a way to provide this tracking. At a high level, the client
provides a map of room ID to room list parameters and the server then returns the response in the
same format as the sliding window API, just without the operations/indexes.

To track a room `!sub1:bar`, the client would send the following request:
```json5
{
  "room_subscriptions": { // sticky

>       "!sub1:bar": { // sticky

Reviewer 20: What should happen if you try to subscribe to a room you are not part of (or not allowed)?

The easy option would be to ignore it but then clients have no feedback about what they did wrong.

Perhaps entries in the `rooms` map could also have an `errcode`/`error` field.

Reviewer 20: Conversation continued in https://github.com/matrix-org/matrix-spec-proposals/pull/4186#discussion_r1767582765

   For example, if the malicious user can engineer which hero's display names are used when calculating
   the room name (say by joining/leaving fake accounts) then it's possible for those names to advertise
   spam or spell out offensive words.

This API presents new ways for clients to request complex operations which runs the risk of denial
of service attacks:
 - Complex or pathological filter/sort options (especially via extensions) may degrade performance
   on the server and client. This may affect other users on the server.
 - Excessively long lists, list keys, ranges, etc. Some limits are specified in this MSC to mitigate against this.
 - Excessive amounts of concurrent connections could consume large amounts of memory on the server for a single device.
   It is recommended that servers limit the number of concurrent connections to 5, and expire the oldest connection
   first.

# Unstable prefix

Whilst this in MSC review the HTTP path will be `/_matrix/client/unstable/org.matrix.msc3575/sync`
with the intention of this eventually becoming (confusingly) `/_matrix/client/v4/sync`. As this is a

> brand new endpoint, no other keys or fields need prefixing.

Reviewer 22: Currently there is no good way for a client to detect presence of a server's support for this MSC unless they are supporting it via proxy (detected via well-known).

If a client tries to make a request via `POST` to the above sync endpoint and the environment is actually running a proxy on the same subdomain then you end up losing `to_device` messages as the proxy starts a new legacy sync stream against the same access token.

Can we please add an entry to https://spec.matrix.org/v1.10/client-server-api/#get_matrixclientversions `unstable_features` to enable the server to express that it has native support.

Workarounds I tried:

OPTIONS request: 200 from the proxy, 204 from Synapse itself
HEAD request: 405 from the proxy, 404 from Synapse itself

Reviewer 4: In lieu of the MSC suggesting something itself, I am assuming `org.matrix.msc3575` will be used.

Homeservers can advertise support for a sliding sync proxy by adding the following to their
`/.well-known/matrix/client` config:
```json
{
    "org.matrix.msc3575.proxy": {

>         "url": "https://slidingsync.proxy.url.here"

Reviewer 23: Can this URL include a path, or should it be a base domain only?

Related issue : https://github.com/element-hq/element-web/issues/27496

Reviewer 24: The original definition of "URL" probably goes back to some RFC from decades ago. I have not looked into it and cannot quote it.

That said, I believe a URL, by definition, contains a path component. The fact that the path component of this `url` field is ignored is quite strange. Ignoring an `#anchor`, etc., - sure, but.. ignoring the path is surprising.

Reviewer 25: I'd say this MSC would have to clarify if `"org.matrix.msc3575.proxy".url` acts like `"m.homeserver".base_url`, for example (as far as I can see the spec doesn't include a formal definition for what a "base url" is either and in the [web url spec](https://url.spec.whatwg.org/#concept-base-url) it's more of an implementation detail as well, although its meaning could be derived form one of the algorithms defined there).

That being said, from what I can see `org.matrix.msc3575.proxy` is only relevant until sync v3 is stable. At that point homeservers are expected to have native implementations of sync v3 (or a proxy would have to hook into the `/_matrix/client/v4/sync` endpoint directly)

Reviewer 26: FWIW, the ElementX implementation honours the paths in the URL

Reviewer 22: The section right above that which is highlighted explains an absolute URL path rather than a relative one

# MSC3575: Client syncing with sliding windows (aka Sync v3)

This MSC outlines a replacement for the CS API endpoint `/sync`.

The current `/sync` endpoint scales badly as the number of rooms on an account increases. It scales
badly because all rooms are returned to the client, and clients cannot opt-out of a large amount of
extraneous data such as receipts. On large accounts with thousands of rooms, the initial sync

> operation can take minutes to perform. This significantly delays the initial login to Matrix clients.

Reviewer 2: ```suggestion
operation can take minutes to perform. This significantly delays the initial login to Matrix clients,
and also makes incremental sync very heavy when resuming after any significant pause in usage.
```

## Goals

Any improved `/sync` mechanism had a number of goals:
 - Sync time should be independent of the number of rooms you are in.
 - Time from launch to confident usability should be as low as possible.
 - Time from login on existing accounts to usability should be as low as possible.
 - Bandwidth should be minimised.
 - Support lazy-loading of things like read receipts (and avoid sending unnecessary data to the client)
 - Support informing the client when room state changes from under it, due to state resolution.
 - Clients should be able to work correctly without ever syncing in the full set of rooms they’re in.
 - Don’t incremental sync rooms you don’t care about.
 - Combining uploaded filters with ad-hoc filter parameters (which isn’t possible with sync v2 today)
 - Servers should not need to store all past since tokens. If a since token has been discarded we should gracefully degrade to initial sync.

>  - Ability to filter by space.

Reviewer 2: we should probably retcon that we were aiming also to support filtering by room name....

These goals shaped the design of this proposal. 

## Proposal

At a high level, the proposal introduces a way for clients to filter and sort the rooms they are
joined to and then request a subset of the resulting list of rooms rather than the entire room list.
```
         All joined rooms on user's account
Q W E R T Y U I O P L K J H G F D S A Z X C V B N M
\                                                 /
 \                                               /
  \      Subset of rooms matched by filters     /
   Q W E R T Y U I O P L K J H G F D S A Z X C V
                       |
   A C D E F G H I J K L O P Q R S T U V W X Y Z     Rooms sorted by name (or by recency, etc)
   |_______|
       |

   A C D E F                                         first 5 rooms requested
```
It also introduces a number of new concepts which are explained in more detail later on:

>  - Core API: The minimumal API to be sync v3 compatible.

Reviewer 2: ```suggestion
 - Core API: The minimal API to be sync v3 compatible.
```

        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
        {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
        ],
        "notification_count": 1,
        "highlight_count": 0
    }
  }

  // Extensions API
  "extensions": {}
}
```
These fields and their interactions are explained in the next few sections. This forms the core of
the API. Additional data can be returned via "extensions". 

#### Connections and streaming data

At a high level, the syncing mechanism creates a "connection" to the server to allow the
bi-directional exchange of JSON objects. This mechanism is ideally suited for WebSockets, but more
difficult to do for HTTP long-polling.

For the long-polling use case, this proposal includes an opaque token that is very similar to
`/sync` v2's `since` query parameter. This is called `pos` and represents the position in the stream
the client is currently at. Unlike `/sync` v2, this token is ephemeral and can be invalidated at any
time. When a client first connects to the server, no `pos` is specified. Also unlike `/sync` v2, this

> token cannot be used with other APIs such as `/messages` or `/keys/changes`.

Reviewer 2: "how do we sync /messages to the current timeline then?", asked the naive reader...

Author: No, it wasn't. /sync v1 included the start/end tokens for use with `/messages` - https://spec.matrix.org/latest/client-server-api/#get_matrixclientv3initialsync

Reviewer 2: okay, but still, how do you correlate with /messages in v3?

Reviewer 2: v2 sync has a `prev_batch` token; how does the same work in v3?

Author: There is a `prev_batch` for each room timeline, but they are explicitly for use with `/messages`, and are separate to the sliding sync `?pos=`.

  | ---------------------> |   data=[A,B], pos=2
  | <--data=[A,B], pos=2-- |
  |                        |   data=[C], pos=3  (new event arrives)
  | -----pos=2-----------> | 
  |   X--data=[C], pos=3-- |  Response is lost
  |                        |
  |                        | data=[C,D], pos=4  (another new event arrives)
  | -----pos=2-----------> | 
  | <----data=[C], pos=3-- | Server CANNOT send data=[C,D] pos=4, it MUST send the previous response
```
Failure to do this will result in duplicate data being sent to the client.

#### Sticky request parameters

Request parameters can be "sticky". This means that their value is remembered across multiple requests.
The lifetime of sticky request parameters are tied to a sync connection. When the connection is lost,
the request parameters are lost with it. This feature exists to allow clients to configure the sync
stream in a bandwidth-efficient way. For example, if all keys were sticky:
```
Client                         Server
  | ------{ "foo": "bar" }------> |  {"foo":"bar"}
  | <-------HTTP 200 OK---------- |
  | ------{ "baz": "quuz" }-----> | {"foo":"bar","baz":"quuz"}
  | <-------HTTP 200 OK---------- |
```

> For complex nested data, APIs which include sticky parameters MUST indicate every sticky field to

Reviewer 2: but how do you indicate that a parameter is sticky?

               | ------{ "baz": "quuz" }-----> | {"foo":"bar","baz":"quuz"}
               |      X--HTTP 200 OK---------- |
               | ------{ "baz": "quuz" }-----> | {"foo":"bar","baz":"quuz"}
               |      X--HTTP 200 OK---------- |
               | ------{ "baz": "quuz" }-----> | {"foo":"bar","baz":"quuz"}
               | <-------HTTP 200 OK---------- |
```
For this reason, some request parameters are not suitable to be made "sticky". These include parameters
which are extremely dynamic in nature, such as list ranges.

#### Room List parameters

One or more room lists can be requested in sync v3 like so:
```js
{
  "lists": [
    {
      // Sliding window ranges, see the Sliding Window API for more information.
      "rooms": [ [0,99] ],
      // Sticky. List sort order. See Sliding Window API for more information.
      "sort": [ "by_notification_count", "by_recency", "by_name" ],

      // Sticky. Required state for each room returned. An array of event type and state key tuples.
      // Note that elements of this array are NOT sticky so they must be specified in full when they
      // are changed.
      "required_state": [

>         // Request the join rules event. Note that the empty string is required here.

Reviewer 2: ```suggestion
        // Request the join rules event. Note that the empty string is required here to match the event's blank state_key.
```

    {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!bar:example.com", "content":{"via":["example.com"]}},
    {"sender":"@alice:example.com","type":"m.space.child", "state_key":"!baz:example.com", "content":{"via":["example.com"]}}
  ],
  // Last event is most recent. Max timeline_limit events.
  "timeline": [
    {"sender":"@alice:example.com","type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"invite"}},
    {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"A"}},
    {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"B"}},
    {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"C"}},
    {"sender":"@alice:example.com","type":"m.room.message", "content":{"body":"D"}},
  ],
  "notification_count": 54, // same as sync v2
  "highlight_count": 3      // same as sync v2
}
```

_Rationale: The room name and counts are required for display on the UI. They are calculated server
side because they are required for sort operations on lists. The `required_state` is controversially
the **current state** which breaks from sync v2 which has the `state` be "the state before the start

> of the timeline". The rationale for this was event duplication and the fact that clients would have

Reviewer 2: i think the rationale for sync v2's behaviour was actually that clients render timelines from old to new (given that's the direction time flows, and new events need to be appended).  Therefore when you get a new batch of messages, you need to know what the state is at the beginning of that batch in order to render the subsequent messages correctly.  Forcing the client developer to 'rewind' state every time it gets a batch of messages in order to then play it forward again (or apply it in reverse) is a needlessly awkward contortion for them to jump through, and it's avoidable.

I don't remember or follow the rationale about event duplication?

to rewind state to work out historical display names. Clients who show historical display names
already need to rewind state by inspecting the `prev_content` of an event to display text like
"@alice changed their name from Alice to Alice2". Event duplication may be
reduced using Event ID -> Event maps in the response, though in practice this duplication does not
happen frequently. The benefit for returning the current state is that servers can cache the latest
state to return the response more quickly, without being forced to rewind this state (as clients will

> need to do) or worse, do an expensive database access to request the state before an event._

Reviewer 2: I'm not sure I follow why it's harder for a server to cache the state at the beginning of a batch than at the end of a batch?

Author: Because the state at the beginning of a batch is variable. Alice, Bob and Charlie may all be at different points in the stream for the same room. To "cache" this then you either need to cache every possible state or track every single position for each client for each room, both of which are impractical.

       DM rooms will have the room name set to the name of the _other user_. This is the value that
       will be returned in the `name` field for the room but is NOT the value that the server should
       perform sort operations on. See following steps.
     * Remove any of the following characters from the beginning/end of the calculated name: `#!()):_@`.
       This ensures things like canonical aliases display in roughly the right locations rather than
       at the start as it starts with `#`.
     * Lower-case the result by unicode. This ensures `Matrix` and `matrix` sort in the same locations.
     * Perform sort operations on this 'canonicalised' name.

_Rationale: The sort operations are restrictive and limited in scope on purpose. Alternatives such
as arbitrary or more expansive sort orders like [RFC4790](https://datatracker.ietf.org/doc/html/rfc4790)
were decided against as it would A) force servers to support nonsensical and potentially expensive
operations and B) not produce the best sort order for specific use cases in Matrix such as alias handling._

#### Room Subscription API
TODO

#### Bandwidth optimisations for persistent clients
TODO

> #### Extensions

Reviewer 2: might make sense to split this and the subsequent extensions into separate MSCs for digestibility

```js
{
    "extensions": {
        "name_of_extension": { // sticky
            "enabled": true, // sticky
            "extension_arg": "value",
            "extension_arg_2": true
        }
    }
}
```
Extensions MUST have an `enabled` flag which defaults to `false`. If a client sends an unknown extension
name, the server MUST ignore it (or else backwards compatibility between servers is broken when a newer
client tries to communicate with an older server). Extension args may or may not be sticky, it
depends on the extension.

Extensions can leverage the data from the core API, notably which rooms are currently inside sliding
windows as well as which rooms are explicitly subscribed to.

### Extensions
#### To Device Messaging
 - Extension name: `to_device`
 - Args:
   * `limit` (Sticky): The max number of events to return per sync response.
   * `since`: The token returned in the `next_batch` section of this extension, or blank if this is the first time.
#### End-to-End Encryption
 - Extension name: `e2ee`

#### Receipts
TODO
#### Typing Notifications
TODO
#### Presence
TODO
#### Account Data
TODO

## Potential issues
TODO

## Alternatives
TODO

> ## Security considerations

Reviewer 2: - incompatibility with encrypted room names, topics and user displaynames - given we need to be able to sort lexicographically and filter based on room name.

client that the response is now an initial sync. For clarity, `initial: true` MUST also be set when
there is no `pos` value provided. When there is a valid `pos`, this flag MUST be omitted (sending
`initial: false` is wasteful).

A response for a given `pos` must be idempotent to account for packet loss. For example:
```
Client                  Server 
  | ---------------------> |   data=[A,B], pos=2
  | <--data=[A,B], pos=2-- |
  |                        |   data=[C], pos=3  (new event arrives)
  | -----pos=2-----------> | 
  |   X--data=[C], pos=3-- |  Response is lost
  |                        |
  |                        | data=[C,D], pos=4  (another new event arrives)
  | -----pos=2-----------> | 
  | <----data=[C], pos=3-- | Server CANNOT send data=[C,D] pos=4, it MUST send the previous response
```
Failure to do this will result in duplicate data being sent to the client.

### Sticky request parameters

Request parameters can be "sticky". This means that their value is remembered across multiple requests.
Clients cannot choose which parameters are sticky, the API defines which parameters are sticky.

> The lifetime of sticky request parameters are tied to a sync connection. When the connection is lost,

Reviewer 22: For the long-polling use case the `pos` token is ephemeral and can be invalidated by the server at any time, thus ending the sync connection and beginning a new one. This sounds like it'd lose any sticky request parameters and thus return an initial sync which does not have the data the client may be expecting due to lost sticky request parameters.

Also given that most HTTP2 sockets/sessions will be terminated at a load balancer, how will servers even know when the connection is closed?

Author: > This sounds like it'd lose any sticky request parameters and thus return an initial sync which does not have the data the client may be expecting due to lost sticky request parameters.

This is a good point. We might need to force a RTT to ensure clients send up their request parameters in full again.

> Also given that most HTTP2 sockets/sessions will be terminated at a load balancer, how will servers even know when the connection is closed?

Connection != TCP connection here, so load balancers aren't a concern. The concept of a sync connection is just the presence of `pos` for a specific access token. `pos`'s absence indicates a new connection and the closure of any currently active connection for that access token.

      "rooms": [ [0,99] ],
      // Sticky. List sort order. See Sliding Window API for more information.
      "sort": [ "by_notification_count", "by_recency", "by_name" ],

      // Sticky. Required state for each room returned. An array of event type and state key tuples.
      // Note that elements of this array are NOT sticky so they must be specified in full when they
      // are changed.
      "required_state": [
        // Request the join rules event. Note that the empty string is required here to match the event's blank state_key.
        ["m.room.join_rules", ""],
        ["m.room.history_visibility", ""],
        // Request all `m.room.member` state events.
        // The * is a special sentinel value meaning 'all keys'.
        // Note that `*` is NOT a generic glob function. You cannot specify `foo*` to pull in keys
        // like `food` and `foobar`. In this case, the * is treated as a literal *.
        ["m.room.member", "*"]
        // Request all state events.
        ["*", "*"]
      ],
      // Sticky. The maximum number of timeline events to return per response.
      "timeline_limit": 10,
      // Sticky. Filters to apply to the list before sorting.

>       "filters": {

Reviewer 22: how would things like Element's Home, Favourites, Orphans "meta spaces" work with the given filters?

Author: In short, I don't know. I'm reluctant to include specific filters and functionality for whatever the current Element UI/UX is because:
 - Element isn't the only client.
 - The UI/UX is in flux, so what is here today may be gone tomorrow which would suck if sync v3 has explicit filters/options for esoteric options which no longer exist.

These points apply to many other Matrix clients as well. An earlier design for Sync v3 for example did include support for Favourites (room tags) but that was explicitly shut down as they really should be personal spaces or some such. I expect Element/Dittochat/Fluffychat/Hydrogen folks to push back on bits they care about, and then we can thrash out what needs to be added to make that UI/UX possible in sync v3. However, when that is taken to the extreme we will inevitably end up with freeform sorting and filtering so it's unfortunately my position to try to maintain some semblance of consistency and simplicity with the resulting API. It would be a major own goal to end up with a slow API because of freeform sorting/filtering which was added to appease everyone.

Reviewer 22: Sure but then how would a client like EW make use of sync v3, it'd need to load all rooms to be able to populate the `Home` space given it has to resolve orphans, so first paint useful paint for logins of huge accounts would still be slow as heck

Author: EW is a set of constantly moving goalposts, so realistically sync v3 would probably be used in more simple/stable (UI/UX wise) clients like Hydrogen first, as well as a long tail of smaller clients which don't have complex UIs but want the speed/bandwidth savings of sync v3.  When EW's UI/UX stops churning then that would be a good time to see what additions are necessary for it to work with sync v3. Obviously it isn't acceptable to make EW pull in all rooms just to populate orphans (so some sort of filter would work here, but it's less clear to know exactly what this filter should be until EW settles on a final UI/UX for spaces).

Reviewer 6: 1. Having filter for rooms not in any space would be cool idea.
2. Favourites could be handled by element using subscribes.
3. Orphans? Is this archive thing? Are those even updated in sync?

Reviewer 27: perhaps `null` in the spaces filter could mean "rooms that aren't in any space"

        // All fields below are Sticky.
        // All fields are applied with AND operators, hence if is_dm:true and is_encrypted:true
        // then only Encrypted DM rooms will be returned. The absence of fields implies no filter
        // on that criteria: it does NOT imply 'false'.

        // Flag which only returns rooms present (or not) in the DM section of account data.
        // If unset, both DM rooms and non-DM rooms are returned. If false, only non-DM rooms
        // are returned. If true, only DM rooms are returned.
        "is_dm": true,
        // A list of spaces which target rooms must be a part of. For every joined room for this

>         // user, ensure that there is a parent space event which is in this list. If unset, all

Reviewer 22: Element clients show invited rooms as belonging to the space too

Author: Thanks, I'll update this to include invited|joined, and clients can then set `is_invite: false` if they don't want to see these rooms.

        // rooms are included. Servers MUST NOT navigate subspaces. It is up to the client to
        // give a complete list of spaces to navigate. Only rooms directly in these spaces will be

>         // returned.

Reviewer 22: So clients would still need to sync all spaces, which would grow linearly with number of spaces they are in, which seems to go against one of the core goals of this MSC

> Sync time should be independent of the number of rooms you are in.

Author: I don't follow your logic here. If a user is present in 5 spaces for example, you surely wouldn't want to be viewing the room list for each and every space at once? Even more complex space hierarchies often hide rooms behind drop-down chevrons and the like, so the act of clicking on that would change the filters. I expect clients to be viewing 1 top-level space at any one time (which then may have N subspaces which may be collapsed or expanded) which fits this API design reasonably well.

Reviewer 22: No but then lets say I want to show the room list for Element Corp, I'd need to know all spaces I'm in to know which spaces make up Element Corp so that then I can pass them to this filter to get the rooms from those spaces, but I don't see an `is_space` filter to even just grab the spaces. 

Unless this wants me to walk the space breadth-first, having to build up this filter as I find more sub-spaces therein.

Filtering on a Space in Element clients includes `(?:grand-)*children`

Reviewer 22: Also even regardless of the room list, to calculate the Notification Badges to show on the Space Panel in Element I'd need to have all the rooms there-in synced up so I can calculate accurate (E2EE included) badge counts

Author: Yes at present you'd need to hit another API to know which spaces a client is part of. I can add in a `is_space` filter pretty easily, but this feels like it's stepping on the toes of the spaces hierarchy endpoint?

Badge counts for spaces was explicitly punted by @ara4n on the basis that the UI/UX was in flux - if they actually are summing all the rooms notif/highlight counts and are tracked in real-time that will add a huge amount of traffic to `/sync` for very little gain (badges changing from 78 to 79 unread messages for example). It's a huge amount of traffic because  literally _any_ event in those rooms will cause traffic to come down `/sync` which simply won't fly for low bandwidth clients even without the event content itself. Having a grey/red dot depending on the presence of unread notifs/mentions respectively seems like a decent trade-off bandwidth wise.

As for accurate E2EE badge counts, you almost certainly _don't_ want to do this as per this MSC mentioning:
 > Grab all the missed messages and see if any of them are highlights. Denial of service risk if there are thousands of messages.

Reviewer 22: > Badge counts for spaces was explicitly punted by @ara4n on the basis that the UI/UX was in flux

From the Spaces team I have certainly not heard any flux in that area whatsoever ftr

>         "spaces": ["!foo:bar", "!bar:baz"],

Reviewer 22: How would this work for DMs in Spaces? The behaviour for DMs in Spaces today is as follows:

Show DMs you have with members of the space (membership = join/invite)
IFF you did not opt-out of this at the space level

![image](https://user-images.githubusercontent.com/2403652/146944163-cfeebacd-c76b-4eaf-96d7-fbebb254f3c9.png)

Author: No idea to be honest. The sync API isn't going to match perfectly whatever Element wants to do. It sounds like you could emulate this by having `is_dm: true` to pull all DMs then client-side filter based on the `m.room.member` state event of the spaces you're interested in showing DMs from.

      "op": "SYNC",
      // The data to put in these positions.
      "rooms": [
        {
          "room_id": "!foo:bar",
          "name": "The calculated room name",
          // Additional response parameters omitted as they are
          // unrelated to the semantics of the sliding window.
          // See previous section on room list parameters.
        },
        // ... 9 more items
      ],
    }
  ],
  // The total number of entries in the list. Index positions match up to the lists in the request.
  "counts": [1337],
}
```

The possible `sort` operations are:
 - `by_recency`: Sort by `origin_server_ts` on the most recently _received_ event in the room. Note
   that due to clock drift over federation it is possible for rooms to re-order such that the most
   recently received event in the entire list does not cause that room to go to index position 0.
 - `by_highlight_count`: Sort by the `highlight_count` for this user in this room, which is the
   number of unread notifications for this room with the highlight flag set. This value is also present
   in sync v2.
 - `by_notification_count`: Sort by the `notification_count` for this user in this room, which is the

>    total number of unread notifications for this room. This value is also present in sync v2.

Reviewer 22: How will this work with E2E rooms where you don't know whether something is a notification before decrypting it, in /sync v2 the notification count had a much lower importance and would be decreased by the client when it confirmed that an encrypted message didn't contain a payload which wants a notification. But now with its importance increasing it may cause bigger impacts.

Author: This is explained in the E2EE section.

>  - `by_name`: Sort by room name lexicographically. This requires servers to implement the

Reviewer 22: This might block any future efforts to encrypt room state e.g name

Author: Indeed, and there is mention of this currently in the security considerations which I will flesh out when the MSC is ready for review. When this was brought up during the design process for Sync v3, the basic conclusion was to punt it for later and hope that some sort of homomorphic encryption might allow us to sort the ciphertext in the future.

```
The update for `{ "other": "value" }` was not applied on the client, but was applied on the server.

With the introduction of buffers and transaction IDs:
```
Updates on receipt of txn ID                                      Updates on receipt of data, every response increments pos by 1, starts at pos=3
                     Client                          Server
                       | -txn=1-{ "foo": "bar" }-pos=3-> |  {"foo":"bar"}  txn=1
{"foo":"bar"}          | <--HTTP 200 OK txn=1,pos=4----- |
                       |                                 |
                      2| -txn=2-{"baz":"quuz"}-pos=4--X  | {"foo":"bar"} txn=1
                       |                                 |
                      3| -txn=3-{"other":"val"}-pos=4--> | {"foo":"bar","other":"val"} txn=3   <-- txn=1 wiped as recv pos=4 in request
                       |    X---HTTP 200 OK txn=3,pos=5- |
                       |                                 |
                      4| -txn=4-{ "k": "v" }----pos=4--> | {"foo":"bar","other":"val","k":"v"}  txn=3,4 <-- txn=3 not wiped as recv old pos in request
{"foo":"bar","k":"v",  | <--HTTP 200 OK txn=3,4,pos=5--- |
 "other":"val"}
```

> If the request parameters have not been modified, then the `txn_id` does not need to be sent.

Reviewer 2: s/sent/changed/ presumably? @kegsay

Author: Well if you don't care about knowing when your params have been applied, sending a txn_id is not really _required_ and you can save a few bytes per request. Saying "changed" implies that you need to send it as-is without modification, and when you change request params you then bump the txn_id to a new value, which isn't what I'm aiming for.

For this reason, some request parameters are not suitable to be made "sticky". These include parameters
which are extremely dynamic in nature, such as list ranges.

### Room List parameters

One or more room lists can be requested in sync v3 like so:

> ```js

Reviewer: suggestion: `json5` could work better with comment-annotated JSON than `js`, for all the usages in this document

      // to return. One unique exception is when you request all state events via ["*", "*"]. When used,
      // all state events are returned by default, and additional entries FILTER OUT the returned set
      // of state events. These additional entries cannot use '*' themselves.
      // For example, ["*", "*"], ["m.room.member", "@alice:example.com"] will _exclude_ every m.room.member
      // event _except_ for @alice:example.com, and include every other state event.
      // In addition, ["*", "*"], ["m.space.child", "*"] is an error, the m.space.child filter is not
      // required as it would have been returned anyway.
      "required_state": [
        // Request the join rules event. Note that the empty string is required here to match
        // the event's blank state_key.
        ["m.room.join_rules", ""],
        ["m.room.history_visibility", ""],
        // Request all `m.space.child` state events.
        // The * is a special sentinel value meaning 'all keys'.
        // Note that `*` is NOT a generic glob function. You cannot specify `foo*` to pull in keys
        // like `food` and `foobar`. In this case, the * is treated as a literal *.

>         ["m.space.child", "*"],

Reviewer 3: ```suggestion
        ["m.space.child", "$ANY"],
```
or
```suggestion
        ["m.space.child", "$ALL"],
```

Seems more consistent?

the **current state** which breaks from sync v2 which has the `state` be "the state before the start
of the timeline". Sync v2's rationale was event duplication (state events can appear in both the
state section and the timeline section if it's the current state) and the fact that clients would have
to rewind state to work out historical display names. Clients who show historical display names
already need to rewind state by inspecting the `prev_content` of an event to display text like
"@alice changed their name from Alice to Alice2". Event duplication may be
reduced using Event ID -> Event maps in the response should this be a concern. The benefit of
returning the current state is that servers can cache the latest state to return the response more
quickly. If, instead, servers returned the state at the start of a timeline block, servers are forced
to either rewind this state (as clients will need to do) or worse, do an expensive database access to
request the state before an event. As clients can be at different points in the stream for a given
room, this would force servers to cache every possible room state. It's not practical for servers to

> cache every single possible earlier state for each room._

Reviewer 2: surely servers need to constantly cache and retrieve earlier states for each room as part of authing events, every time you get an event with a parent that isn't an extremity?

Author: Yes, and this is expensive. Servers can offload this to sharded room server components, and this works well because the work is per-room. Servers become very difficult to optimise if synchotrons/sync servers need to do the same operation (expensive per room operations on a per-user sharded process).

    // the room ID
    "!foo:bar": {
      "name": "The calculated room name",

>       // The MXC URL of the room's avatar, or the string "none" if this room has no prescribed avatar.

Reviewer 18: Because 
- the general design (correct me if I'm wrong @kegsay ) is that `null` is treated exactly the same as an unspecified field
- an omitted avatar field means "no change to the avatar".

>       // The MXC URL of the room's avatar, or the string "none" if this room has no prescribed avatar.

Reviewer 18: Wording changed in 912621b4.

For the first of these issues, the sliding sync API exposes a "room subscription" API. For the second issue,
the sliding sync API exposes a "notifications" API.

#### Requesting all rooms

Sometimes clients may not wish to deal with sliding windows, and instead get all rooms on the user's account.
For example, if your client is a bot or an application service, having sliding windows just adds extra complexity.
To aid these use cases, any list can omit the `ranges` key and add a new sticky key at the same level:

> `slow_get_all_rooms: true`. If this is set, the `ranges` and `sort` keys are ignored and all rooms which match the list

Reviewer: Why does this have to be prepended with `slow_`?

Author: This was a conscious choice so developers know that setting this flag introduces a performance penalty. This is the least obnoxious way to do it, with the inspiration coming from React-style `__dangerouslySetInnerHtml`.

   that due to clock drift over federation it is possible for rooms to re-order such that the most
   recently received event in the entire list does not cause that room to go to index position 0.
   The highest `origin_server_ts` value comes first in the list.
 - `by_notification_level`: Sort based on the presence of non-zero values for `highlight_count` and
   `notification_count`. Rooms with a `highlight_count` > 0 come first, followed by rooms with a

>    `notification_count` > 0, followed by all other rooms. Rooms are not sorted within each level:

Reviewer 11: If no other sorting giving, I'd expect them to sort by the _number_ within each ... so a room with `highlight_count = 4` comes before `highlight_count = 3`, etc ...

Author: This is semantically not useful, and would then break the ability to do 2nd order sort options (you couldn't then say `by_recency` as there would be no ties, or no ties at a sensible point rather as you'd get ties when the counts are the same).

_Rationale: The name 'extensions' is inspired by the spec itself which refers to "Extensions to /sync"
multiple times. These additional bits of data are all generally outside the scope of the core room
graph and room list so are well-placed for being treated separately. Furthermore, it is possible to
make a meaningful client which only supports the core API and no extensions, as the core controls
the room list and ability to receive events and state in a room. For clients which don't do E2EE and
don't handle presence/typing/receipts/other metadata, they can simply work with this MSC alone and
in full._

# Potential issues
- Scope.
- Flexibility with sorting/filtering and client demands.

# Alternatives
- Do nothing, keep using sync v2 and try to make it go fast.
- Factor out some busy bits of sync v2 e.g receipts but still keep returning all rooms (no paginated sync)

# Security considerations
- room sub auth check
- spaces auth check
- history visibility for timeline_limit
- incompatibility with encrypted room names, topics and user displaynames - given we need to be able
  to sort lexicographically and filter based on room name.

# Unstable prefix

> - Annoyingly probably `/v4/sync`

Reviewer: Why not prefixed with `org.matrix.msc3575`? Rationale is missing.

with the intention of this eventually becoming (confusingly) `/_matrix/client/v4/sync`. As this is a
brand new endpoint, no other keys or fields need prefixing.

Homeservers can advertise support for a sliding sync proxy by adding the following to their
`/.well-known/matrix/client` config:
```json
{
    "org.matrix.msc3575.proxy": {
        "url": "https://slidingsync.proxy.url.here"
    }
}
```
This allows servers to declare an "official" trusted proxy, rather than using other URLs which may be
run by malicious actors who want to steal the access token for users.

> 

Reviewer 11: Should this there mandate that _clients must use_ this as the target if they find it?

Reviewer 22: Why does a server need advertise a proxy in this way? Why not just make it offer it as if the server natively supported the feature, i.e. via a reverse proxy? Given some implementations may internalise the proxy by wrapping it in a worker (e.g. Dendrite) why not make the access consistent
